<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Overview of the immunogenetr package</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="immunogenetr_overview_files/libs/clipboard/clipboard.min.js"></script>
<script src="immunogenetr_overview_files/libs/quarto-html/quarto.js"></script>
<script src="immunogenetr_overview_files/libs/quarto-html/popper.min.js"></script>
<script src="immunogenetr_overview_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="immunogenetr_overview_files/libs/quarto-html/anchor.min.js"></script>
<link href="immunogenetr_overview_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="immunogenetr_overview_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="immunogenetr_overview_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="immunogenetr_overview_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="immunogenetr_overview_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Overview of the immunogenetr package</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="immunogenetr-an-hla-data-wrangling-package-for-r" class="level2">
<h2 class="anchored" data-anchor-id="immunogenetr-an-hla-data-wrangling-package-for-r">immunogenetr: an HLA data wrangling package for R</h2>
<p>Immunogenetics is the study of the major histocompatbility complex (MHC). The human MHC is called HLA, which has serologic and molecular nomenclatures define by the World Health Organization (WHO; references). Beyond the WHO-defined nomenclature, the genotype list string (GL string) is a standard for encoding HLA genotypes, with standard delimiters for encoding multiple loci, gene copies, ambiguities, etc. (reference). HLA genotype data from primary sources, including clinical laboratory information systems or research databases, are often stored in a variety of formats. The goal of the immunogenetr package is to provide a set of functions for coercing HLA data into a standard format: the GL string. Other functions then perform common tasks, such as mismatch determination, or separation of loci to individual columns, from the GL string. Various other functions allow for further wrangling of HLA data. This package relies heavily on functions from, and adheres to concepts of, the tidyverse (reference). It is envisioned that immunogenetr will allow for the rapid and accurate wrangling of HLA data for downstream analysis.</p>
<p>This document will provide an overview of the functions and their intended use.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The current package can be installed from github</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">"k96nb01/immunogenetr_package"</span>, <span class="at">force =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="importing-wild-caught-data" class="level3">
<h3 class="anchored" data-anchor-id="importing-wild-caught-data">Importing wild-caught data</h3>
<p>One of the most common formats for HLA genotype data is in spreadsheet-like table where each row represents a case and each column a single allele of one of the HLA loci. See below for examples of this type of wild-caught data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::between()     masks data.table::between()
✖ dplyr::filter()      masks stats::filter()
✖ dplyr::first()       masks data.table::first()
✖ lubridate::hour()    masks data.table::hour()
✖ lubridate::isoweek() masks data.table::isoweek()
✖ dplyr::lag()         masks stats::lag()
✖ dplyr::last()        masks data.table::last()
✖ lubridate::mday()    masks data.table::mday()
✖ lubridate::minute()  masks data.table::minute()
✖ lubridate::month()   masks data.table::month()
✖ lubridate::quarter() masks data.table::quarter()
✖ lubridate::second()  masks data.table::second()
✖ purrr::transpose()   masks data.table::transpose()
✖ lubridate::wday()    masks data.table::wday()
✖ lubridate::week()    masks data.table::week()
✖ lubridate::yday()    masks data.table::yday()
✖ lubridate::year()    masks data.table::year()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"tests/test1.csv"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>test4 <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"tests/test4.csv"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(test1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   patient      A1      A2      C1      C2      B1      B2   DRB345_1
1        1 A*24:02 A*29:02 C*07:04 C*16:01 B*44:02 B*44:03 DRB5*01:01
2        2 A*02:01 A*11:05 C*07:01 C*07:02 B*07:02 B*08:01 DRB3*01:01
3        3 A*02:01 A*26:18 C*02:02 C*03:04 B*27:05 B*54:01 DRB3*02:02
4        4 A*29:02 A*30:02 C*06:02 C*07:01 B*08:01 B*13:02 DRB4*01:03
5        5 A*02:05 A*24:02 C*07:18 C*12:03 B*35:03 B*58:01 DRB3*02:02
6        6 A*01:01 A*24:02 C*07:01 C*14:02 B*49:01 B*51:01 DRB3*03:01
7        7 A*03:01 A*03:01 C*03:03 C*16:01 B*15:01 B*51:01 DRB4*01:01
8        8 A*01:01 A*32:01 C*06:02 C*07:02 B*08:01 B*37:01 DRB3*02:02
9        9 A*03:01 A*30:01 C*07:02 C*12:03 B*07:02 B*38:01 DRB3*01:01
10      10 A*02:05 A*11:01 C*07:18 C*16:02 B*51:01 B*58:01 DRB3*03:01
     DRB345_2     DRB1_1     DRB1_2     DQA1_1     DQA1_2     DQB1_1     DQB1_2
1  DRB5*01:01 DRB1*15:01 DRB1*15:01 DQA1*01:02 DQA1*01:02 DQB1*06:02 DQB1*06:02
2  DRB4*01:03 DRB1*03:01 DRB1*04:01 DQA1*03:03 DQA1*05:01 DQB1*02:01 DQB1*03:01
3  DRB4*01:03 DRB1*04:04 DRB1*14:54 DQA1*01:04 DQA1*03:01 DQB1*03:02 DQB1*05:02
4  DRB4*01:03 DRB1*04:01 DRB1*07:01 DQA1*02:01 DQA1*03:01 DQB1*02:02 DQB1*03:02
5  DRB3*02:02 DRB1*03:01 DRB1*14:54 DQA1*01:04 DQA1*05:01 DQB1*02:01 DQB1*05:03
6   DRBX*NNNN DRB1*08:01 DRB1*13:02 DQA1*01:02 DQA1*04:01 DQB1*04:02 DQB1*06:04
7   DRBX*NNNN DRB1*01:01 DRB1*07:01 DQA1*01:01 DQA1*02:01 DQB1*02:02 DQB1*05:01
8  DRB5*01:01 DRB1*03:01 DRB1*15:01 DQA1*01:02 DQA1*05:01 DQB1*02:01 DQB1*06:02
9  DRB5*01:01 DRB1*03:01 DRB1*15:01 DQA1*01:02 DQA1*05:01 DQB1*02:01 DQB1*06:02
10 DRB5*01:01 DRB1*13:02 DRB1*15:01 DQA1*01:02 DQA1*01:03 DQB1*06:01 DQB1*06:09
       DPA1_1     DPA1_2     DPB1_1      DPB1_2
1  DPA1*01:03 DPA1*01:03 DPB1*03:01  DPB1*04:01
2  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
3  DPA1*01:03 DPA1*02:02 DPB1*02:01  DPB1*05:01
4  DPA1*01:03 DPA1*02:01 DPB1*01:01  DPB1*16:01
5  DPA1*01:03 DPA1*02:01 DPB1*10:01 DPB1*124:01
6  DPA1*01:03 DPA1*01:04 DPB1*04:01  DPB1*15:01
7  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
8  DPA1*01:03 DPA1*02:01 DPB1*04:01  DPB1*14:01
9  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
10 DPA1*01:03 DPA1*01:03 DPB1*02:01 DPB1*104:01</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(test4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   patient mA1Cd.recipient mA2Cd.recipient mB1Cd.recipient mB2Cd.recipient
1        1           24:02           02:01           40:02           40:01
2        2           03:01           74:01           53:01           57:03
3        3           11:01           32:01           52:01                
4        4           03:01           30:02           14:02                
5        5           01:01           24:02           07:02                
6        6           02:01           30:02           39:11           41:01
7        7           02:17           32:01           40:02                
8        8           11:01           23:01           15:17           44:03
9        9           03:01           68:02           08:01           15:03
10      10           01:01           02:01           08:01           07:02
   mC1Cd.recipient mC2Cd.recipient mDRB11Cd.recipient mDRB12Cd.recipient
1            03:04                              10:01              15:01
2            04:01           07:01              11:01              16:02
3            12:02                              15:02                   
4            08:02                              01:02              13:02
5            07:02                              01:01              15:01
6            07:02           17:01              03:01              04:05
7            02:02           03:05              04:11              13:01
8            07:01           16:01              07:01                   
9            02:10           03:04              13:04              07:01
10           07:01           05:01              03:01              04:01
   mDRB31cd.recipient mDRB32cd.recipient mDRB41cd.recipient mDRB42cd.recipient
1                                                                             
2               02:02                                                         
3                                                                             
4                                  03:01                                      
5                                                                             
6               02:02                                                    01:03
7                                  02:02              01:03                   
8                                                     01:01              01:03
9               02:02                                                    01:03
10              01:01                                                    01:03
   mDRB51cd.recipient mDRB52cd.recipient mDQA11Cd.recipient mDQA12Cd.recipient
1                                  01:01              01:05              01:02
2                                  02:21              01:02              05:05
3               01:02                                 01:03                   
4                                                     01:01              01:02
5                                  01:01              01:01              01:02
6                                                     05:01              03:03
7                                                     03:01              01:03
8                                                     02:01                   
9                                                     02:01              05:05
10                                                    05:01              03:03
   mDQB11cd.recipient mDQB12cd.recipient mDPA11Cd.recipient mDPA12Cd.recipient
1               05:01              06:02              01:03                   
2               03:19              05:02              02:01              02:02
3               06:01                                 01:03              02:07
4               05:01              06:09              02:01                   
5               05:01              06:02              01:03              02:01
6               02:01              02:02              01:03                   
7               03:02              06:03              01:03                   
8               02:02                                 01:03              02:01
9               02:02              03:19              01:03              02:02
10              02:01              03:01              01:03                   
   mDPB11cd.recipient mDPB12cd.recipient
1               02:01              04:01
2               01:01                   
3               02:01              04:01
4               05:01              17:01
5               04:01              11:01
6               02:01              04:01
7               04:02                   
8               04:01              11:01
9               01:01              02:01
10              04:01              06:01</code></pre>
</div>
</div>
<p>Note that in the first example, the DRB3, DRB4, and DRB5 loci are encoded in two columns, corresponding to one locus, while in the second, each of these loci has two columns for the two possible alleles. In both examples, column names contain “1” and “2” suffixes to indicate gene copies, while the second example has other suffixes and prefixes in the column names of the loci. The <code>HLA_column_to_GLstring</code> function was written to coerce these disparate formats to a standard GL string:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>test1 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">GL_string =</span> <span class="fu">HLA_columns_to_GLstring</span>(., patient, <span class="fu">c</span>(A1<span class="sc">:</span>DPB1_2)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   patient      A1      A2      C1      C2      B1      B2   DRB345_1
1        1 A*24:02 A*29:02 C*07:04 C*16:01 B*44:02 B*44:03 DRB5*01:01
2        2 A*02:01 A*11:05 C*07:01 C*07:02 B*07:02 B*08:01 DRB3*01:01
3        3 A*02:01 A*26:18 C*02:02 C*03:04 B*27:05 B*54:01 DRB3*02:02
4        4 A*29:02 A*30:02 C*06:02 C*07:01 B*08:01 B*13:02 DRB4*01:03
5        5 A*02:05 A*24:02 C*07:18 C*12:03 B*35:03 B*58:01 DRB3*02:02
6        6 A*01:01 A*24:02 C*07:01 C*14:02 B*49:01 B*51:01 DRB3*03:01
7        7 A*03:01 A*03:01 C*03:03 C*16:01 B*15:01 B*51:01 DRB4*01:01
8        8 A*01:01 A*32:01 C*06:02 C*07:02 B*08:01 B*37:01 DRB3*02:02
9        9 A*03:01 A*30:01 C*07:02 C*12:03 B*07:02 B*38:01 DRB3*01:01
10      10 A*02:05 A*11:01 C*07:18 C*16:02 B*51:01 B*58:01 DRB3*03:01
     DRB345_2     DRB1_1     DRB1_2     DQA1_1     DQA1_2     DQB1_1     DQB1_2
1  DRB5*01:01 DRB1*15:01 DRB1*15:01 DQA1*01:02 DQA1*01:02 DQB1*06:02 DQB1*06:02
2  DRB4*01:03 DRB1*03:01 DRB1*04:01 DQA1*03:03 DQA1*05:01 DQB1*02:01 DQB1*03:01
3  DRB4*01:03 DRB1*04:04 DRB1*14:54 DQA1*01:04 DQA1*03:01 DQB1*03:02 DQB1*05:02
4  DRB4*01:03 DRB1*04:01 DRB1*07:01 DQA1*02:01 DQA1*03:01 DQB1*02:02 DQB1*03:02
5  DRB3*02:02 DRB1*03:01 DRB1*14:54 DQA1*01:04 DQA1*05:01 DQB1*02:01 DQB1*05:03
6   DRBX*NNNN DRB1*08:01 DRB1*13:02 DQA1*01:02 DQA1*04:01 DQB1*04:02 DQB1*06:04
7   DRBX*NNNN DRB1*01:01 DRB1*07:01 DQA1*01:01 DQA1*02:01 DQB1*02:02 DQB1*05:01
8  DRB5*01:01 DRB1*03:01 DRB1*15:01 DQA1*01:02 DQA1*05:01 DQB1*02:01 DQB1*06:02
9  DRB5*01:01 DRB1*03:01 DRB1*15:01 DQA1*01:02 DQA1*05:01 DQB1*02:01 DQB1*06:02
10 DRB5*01:01 DRB1*13:02 DRB1*15:01 DQA1*01:02 DQA1*01:03 DQB1*06:01 DQB1*06:09
       DPA1_1     DPA1_2     DPB1_1      DPB1_2
1  DPA1*01:03 DPA1*01:03 DPB1*03:01  DPB1*04:01
2  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
3  DPA1*01:03 DPA1*02:02 DPB1*02:01  DPB1*05:01
4  DPA1*01:03 DPA1*02:01 DPB1*01:01  DPB1*16:01
5  DPA1*01:03 DPA1*02:01 DPB1*10:01 DPB1*124:01
6  DPA1*01:03 DPA1*01:04 DPB1*04:01  DPB1*15:01
7  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
8  DPA1*01:03 DPA1*02:01 DPB1*04:01  DPB1*14:01
9  DPA1*01:03 DPA1*01:03 DPB1*04:01  DPB1*04:01
10 DPA1*01:03 DPA1*01:03 DPB1*02:01 DPB1*104:01
                                                                                                                                                                                                                                                      GL_string
1   HLA-A*24:02+HLA-A*29:02^HLA-C*07:04+HLA-C*16:01^HLA-B*44:02+HLA-B*44:03^HLA-DRB5*01:01+HLA-DRB5*01:01^HLA-DRB1*15:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*01:02^HLA-DQB1*06:02+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*03:01+HLA-DPB1*04:01
2   HLA-A*02:01+HLA-A*11:05^HLA-C*07:01+HLA-C*07:02^HLA-B*07:02+HLA-B*08:01^HLA-DRB3*01:01^HLA-DRB4*01:03^HLA-DRB1*03:01+HLA-DRB1*04:01^HLA-DQA1*03:03+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*03:01^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01
3   HLA-A*02:01+HLA-A*26:18^HLA-C*02:02+HLA-C*03:04^HLA-B*27:05+HLA-B*54:01^HLA-DRB3*02:02^HLA-DRB4*01:03^HLA-DRB1*04:04+HLA-DRB1*14:54^HLA-DQA1*01:04+HLA-DQA1*03:01^HLA-DQB1*03:02+HLA-DQB1*05:02^HLA-DPA1*01:03+HLA-DPA1*02:02^HLA-DPB1*02:01+HLA-DPB1*05:01
4   HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01
5  HLA-A*02:05+HLA-A*24:02^HLA-C*07:18+HLA-C*12:03^HLA-B*35:03+HLA-B*58:01^HLA-DRB3*02:02+HLA-DRB3*02:02^HLA-DRB1*03:01+HLA-DRB1*14:54^HLA-DQA1*01:04+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*05:03^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*10:01+HLA-DPB1*124:01
6                  HLA-A*01:01+HLA-A*24:02^HLA-C*07:01+HLA-C*14:02^HLA-B*49:01+HLA-B*51:01^HLA-DRB3*03:01^HLA-DRB1*08:01+HLA-DRB1*13:02^HLA-DQA1*01:02+HLA-DQA1*04:01^HLA-DQB1*04:02+HLA-DQB1*06:04^HLA-DPA1*01:03+HLA-DPA1*01:04^HLA-DPB1*04:01+HLA-DPB1*15:01
7                  HLA-A*03:01+HLA-A*03:01^HLA-C*03:03+HLA-C*16:01^HLA-B*15:01+HLA-B*51:01^HLA-DRB4*01:01^HLA-DRB1*01:01+HLA-DRB1*07:01^HLA-DQA1*01:01+HLA-DQA1*02:01^HLA-DQB1*02:02+HLA-DQB1*05:01^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01
8   HLA-A*01:01+HLA-A*32:01^HLA-C*06:02+HLA-C*07:02^HLA-B*08:01+HLA-B*37:01^HLA-DRB3*02:02^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*04:01+HLA-DPB1*14:01
9   HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01
10 HLA-A*02:05+HLA-A*11:01^HLA-C*07:18+HLA-C*16:02^HLA-B*51:01+HLA-B*58:01^HLA-DRB3*03:01^HLA-DRB5*01:01^HLA-DRB1*13:02+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*01:03^HLA-DQB1*06:01+HLA-DQB1*06:09^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*02:01+HLA-DPB1*104:01</code></pre>
</div>
</div>
<p>For the second example, the <code>prefix_to_remove</code> and <code>suffix_to_remove</code> arguments ensure the function accurately determines the loci for each column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>test4 <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">GL_string =</span> <span class="fu">HLA_columns_to_GLstring</span>(., patient, <span class="fu">c</span>(mA1Cd.recipient<span class="sc">:</span>mDPB12cd.recipient), <span class="at">prefix_to_remove =</span> <span class="st">"m"</span>, <span class="at">suffix_to_remove =</span> <span class="st">"Cd.recipient"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   patient mA1Cd.recipient mA2Cd.recipient mB1Cd.recipient mB2Cd.recipient
1        1           24:02           02:01           40:02           40:01
2        2           03:01           74:01           53:01           57:03
3        3           11:01           32:01           52:01                
4        4           03:01           30:02           14:02                
5        5           01:01           24:02           07:02                
6        6           02:01           30:02           39:11           41:01
7        7           02:17           32:01           40:02                
8        8           11:01           23:01           15:17           44:03
9        9           03:01           68:02           08:01           15:03
10      10           01:01           02:01           08:01           07:02
   mC1Cd.recipient mC2Cd.recipient mDRB11Cd.recipient mDRB12Cd.recipient
1            03:04                              10:01              15:01
2            04:01           07:01              11:01              16:02
3            12:02                              15:02                   
4            08:02                              01:02              13:02
5            07:02                              01:01              15:01
6            07:02           17:01              03:01              04:05
7            02:02           03:05              04:11              13:01
8            07:01           16:01              07:01                   
9            02:10           03:04              13:04              07:01
10           07:01           05:01              03:01              04:01
   mDRB31cd.recipient mDRB32cd.recipient mDRB41cd.recipient mDRB42cd.recipient
1                                                                             
2               02:02                                                         
3                                                                             
4                                  03:01                                      
5                                                                             
6               02:02                                                    01:03
7                                  02:02              01:03                   
8                                                     01:01              01:03
9               02:02                                                    01:03
10              01:01                                                    01:03
   mDRB51cd.recipient mDRB52cd.recipient mDQA11Cd.recipient mDQA12Cd.recipient
1                                  01:01              01:05              01:02
2                                  02:21              01:02              05:05
3               01:02                                 01:03                   
4                                                     01:01              01:02
5                                  01:01              01:01              01:02
6                                                     05:01              03:03
7                                                     03:01              01:03
8                                                     02:01                   
9                                                     02:01              05:05
10                                                    05:01              03:03
   mDQB11cd.recipient mDQB12cd.recipient mDPA11Cd.recipient mDPA12Cd.recipient
1               05:01              06:02              01:03                   
2               03:19              05:02              02:01              02:02
3               06:01                                 01:03              02:07
4               05:01              06:09              02:01                   
5               05:01              06:02              01:03              02:01
6               02:01              02:02              01:03                   
7               03:02              06:03              01:03                   
8               02:02                                 01:03              02:01
9               02:02              03:19              01:03              02:02
10              02:01              03:01              01:03                   
   mDPB11cd.recipient mDPB12cd.recipient
1               02:01              04:01
2               01:01                   
3               02:01              04:01
4               05:01              17:01
5               04:01              11:01
6               02:01              04:01
7               04:02                   
8               04:01              11:01
9               01:01              02:01
10              04:01              06:01
                                                                                                                                                                                                                                                     GL_string
1                                            HLA-A*24:02+HLA-A*02:01^HLA-B*40:02+HLA-B*40:01^HLA-C*03:04^HLA-DRB1*10:01+HLA-DRB1*15:01^HLA-DRB5*01:01^HLA-DQA1*01:05+HLA-DQA1*01:02^HLA-DQB1*05:01+HLA-DQB1*06:02^HLA-DPA1*01:03^HLA-DPB1*02:01+HLA-DPB1*04:01
2                 HLA-A*03:01+HLA-A*74:01^HLA-B*53:01+HLA-B*57:03^HLA-C*04:01+HLA-C*07:01^HLA-DRB1*11:01+HLA-DRB1*16:02^HLA-DRB3*02:02^HLA-DRB5*02:21^HLA-DQA1*01:02+HLA-DQA1*05:05^HLA-DQB1*03:19+HLA-DQB1*05:02^HLA-DPA1*02:01+HLA-DPA1*02:02^HLA-DPB1*01:01
3                                                                                      HLA-A*11:01+HLA-A*32:01^HLA-B*52:01^HLA-C*12:02^HLA-DRB1*15:02^HLA-DRB5*01:02^HLA-DQA1*01:03^HLA-DQB1*06:01^HLA-DPA1*01:03+HLA-DPA1*02:07^HLA-DPB1*02:01+HLA-DPB1*04:01
4                                                        HLA-A*03:01+HLA-A*30:02^HLA-B*14:02^HLA-C*08:02^HLA-DRB1*01:02+HLA-DRB1*13:02^HLA-DRB3*03:01^HLA-DQA1*01:01+HLA-DQA1*01:02^HLA-DQB1*05:01+HLA-DQB1*06:09^HLA-DPA1*02:01^HLA-DPB1*05:01+HLA-DPB1*17:01
5                                         HLA-A*01:01+HLA-A*24:02^HLA-B*07:02^HLA-C*07:02^HLA-DRB1*01:01+HLA-DRB1*15:01^HLA-DRB5*01:01^HLA-DQA1*01:01+HLA-DQA1*01:02^HLA-DQB1*05:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*04:01+HLA-DPB1*11:01
6                 HLA-A*02:01+HLA-A*30:02^HLA-B*39:11+HLA-B*41:01^HLA-C*07:02+HLA-C*17:01^HLA-DRB1*03:01+HLA-DRB1*04:05^HLA-DRB3*02:02^HLA-DRB4*01:03^HLA-DQA1*05:01+HLA-DQA1*03:03^HLA-DQB1*02:01+HLA-DQB1*02:02^HLA-DPA1*01:03^HLA-DPB1*02:01+HLA-DPB1*04:01
7                                            HLA-A*02:17+HLA-A*32:01^HLA-B*40:02^HLA-C*02:02+HLA-C*03:05^HLA-DRB1*04:11+HLA-DRB1*13:01^HLA-DRB3*02:02^HLA-DRB4*01:03^HLA-DQA1*03:01+HLA-DQA1*01:03^HLA-DQB1*03:02+HLA-DQB1*06:03^HLA-DPA1*01:03^HLA-DPB1*04:02
8                                               HLA-A*11:01+HLA-A*23:01^HLA-B*15:17+HLA-B*44:03^HLA-C*07:01+HLA-C*16:01^HLA-DRB1*07:01^HLA-DRB4*01:01+HLA-DRB4*01:03^HLA-DQA1*02:01^HLA-DQB1*02:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*04:01+HLA-DPB1*11:01
9  HLA-A*03:01+HLA-A*68:02^HLA-B*08:01+HLA-B*15:03^HLA-C*02:10+HLA-C*03:04^HLA-DRB1*13:04+HLA-DRB1*07:01^HLA-DRB3*02:02^HLA-DRB4*01:03^HLA-DQA1*02:01+HLA-DQA1*05:05^HLA-DQB1*02:02+HLA-DQB1*03:19^HLA-DPA1*01:03+HLA-DPA1*02:02^HLA-DPB1*01:01+HLA-DPB1*02:01
10                HLA-A*01:01+HLA-A*02:01^HLA-B*08:01+HLA-B*07:02^HLA-C*07:01+HLA-C*05:01^HLA-DRB1*03:01+HLA-DRB1*04:01^HLA-DRB3*01:01^HLA-DRB4*01:03^HLA-DQA1*05:01+HLA-DQA1*03:03^HLA-DQB1*02:01+HLA-DQB1*03:01^HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*06:01</code></pre>
</div>
</div>
<p>This function also allows for the ability to coerce serologic typing to a GL string. While the GL string was not intended to contain serologic results, the format will accommodate it, and allow for downstream functions to act on serologic data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>(test_a <span class="ot">&lt;-</span> <span class="fu">tribble</span>(</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span>case, <span class="sc">~</span><span class="st">"a1"</span>, <span class="sc">~</span><span class="st">"a2"</span>, <span class="sc">~</span><span class="st">"drw1"</span>, <span class="sc">~</span><span class="st">"drw2"</span>, <span class="sc">~</span><span class="st">"cw1"</span>, <span class="sc">~</span><span class="st">"cw2"</span>,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>, <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"51"</span>, <span class="st">"52"</span>, <span class="st">"1"</span>, <span class="st">"15"</span>,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="dv">2</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"53"</span>, <span class="st">"blank"</span>, <span class="st">"10"</span>, <span class="st">"blank"</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  <span class="dv">3</span>, <span class="st">"2"</span>, <span class="cn">NA</span>, <span class="st">"51"</span>, <span class="st">"51"</span>, <span class="st">"3"</span>, <span class="cn">NA</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  <span class="dv">4</span>, <span class="st">"68"</span>, <span class="cn">NA</span>, <span class="cn">NA</span>, <span class="st">"53"</span>, <span class="st">"5"</span>, <span class="st">"17"</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 7
   case a1    a2    drw1  drw2  cw1   cw2  
  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;
1     1 1     2     51    52    1     15   
2     2 2     3     53    blank 10    blank
3     3 2     &lt;NA&gt;  51    51    3     &lt;NA&gt; 
4     4 68    &lt;NA&gt;  &lt;NA&gt;  53    5     17   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>(test_a <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">GL_string =</span> <span class="fu">HLA_columns_to_GLstring</span>(., case, <span class="fu">c</span>(a1<span class="sc">:</span>cw2), <span class="at">serologic =</span> <span class="cn">TRUE</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 8
   case a1    a2    drw1  drw2  cw1   cw2   GL_string                           
  &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;                               
1     1 1     2     51    52    1     15    HLA-A1+HLA-A2^HLA-DR51+HLA-DR52^HLA…
2     2 2     3     53    blank 10    blank HLA-A2+HLA-A3^HLA-DR53^HLA-Cw10     
3     3 2     &lt;NA&gt;  51    51    3     &lt;NA&gt;  HLA-A2^HLA-DR51+HLA-DR51^HLA-Cw3    
4     4 68    &lt;NA&gt;  &lt;NA&gt;  53    5     17    HLA-A68^HLA-DR53^HLA-Cw5+HLA-Cw17   </code></pre>
</div>
</div>
<p>In addition to making GL strings from data stored in individual columns, I also have a function to read the GL string from an HML file. I haven’t uploaded this yet, though, as I need to get a deidentified HML file that I can upload as a test.</p>
</section>
<section id="working-with-gl-strings" class="level3">
<h3 class="anchored" data-anchor-id="working-with-gl-strings">Working with GL strings</h3>
<p>For working with GL strings, there are functions to parse them. The first would be to separate genes (loci):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>(test <span class="ot">&lt;-</span> <span class="fu">tribble</span>(</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span>case, <span class="sc">~</span>string, </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>, <span class="st">"HLA-A*02:01+HLA-A*29:02^HLA-B*44:03+HLA-B*49:01^HLA-C*07:01+HLA-C*16:01^HLA-DRB1*04:05+HLA-DRB1*07:01^HLA-DRB4*01:01+HLA-DRB4*01:03^HLA-DQA1*02:01+HLA-DQA1*03:03^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*11:01+HLA-DPB1*104:01  "</span>, </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="dv">2</span>, <span class="st">"HLA-A*03:01+HLA-A*68:01^HLA-C*07:01+HLA-C*07:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:02+HLA-DPB1*04:02^HLA-DQA1*01:02+HLA-DQA1*05:05^HLA-DQB1*03:01+HLA-DQB1*05:02^HLA-DRB1*15:01+HLA-DRB1*12:01^HLA-DRB3*02:02^HLA-DRB5*01:01+HLA-DRB5*01:01"</span>, </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="dv">3</span>, <span class="st">"HLA-A*02:01:01^HLA-E*01:02"</span>, </span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="dv">4</span>, <span class="st">"HLA-A*02:01:01+HLA-A*32:01:01^MICA*001:01+MICA*008:01^MICB*001:01"</span>, </span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  ))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 2
   case string                                                                  
  &lt;dbl&gt; &lt;chr&gt;                                                                   
1     1 "HLA-A*02:01+HLA-A*29:02^HLA-B*44:03+HLA-B*49:01^HLA-C*07:01+HLA-C*16:0…
2     2 "HLA-A*03:01+HLA-A*68:01^HLA-C*07:01+HLA-C*07:02^HLA-DPA1*01:03+HLA-DPA…
3     3 "HLA-A*02:01:01^HLA-E*01:02"                                            
4     4 "HLA-A*02:01:01+HLA-A*32:01:01^MICA*001:01+MICA*008:01^MICB*001:01"     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>test <span class="sc">%&gt;%</span> <span class="fu">GLstring_genes</span>(string)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 15
   case HLA_A  HLA_B HLA_C HLA_DRB1 HLA_DRB4 HLA_DQA1 HLA_DQB1 HLA_DPA1 HLA_DPB1
  &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;   
1     1 HLA-A… HLA-… HLA-… HLA-DRB… HLA-DRB… HLA-DQA… HLA-DQB… HLA-DPA… "HLA-DP…
2     2 HLA-A… &lt;NA&gt;  HLA-… HLA-DRB… &lt;NA&gt;     HLA-DQA… HLA-DQB… HLA-DPA… "HLA-DP…
3     3 HLA-A… &lt;NA&gt;  &lt;NA&gt;  &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;   
4     4 HLA-A… &lt;NA&gt;  &lt;NA&gt;  &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;     &lt;NA&gt;      &lt;NA&gt;   
# ℹ 5 more variables: HLA_DRB3 &lt;chr&gt;, HLA_DRB5 &lt;chr&gt;, HLA_E &lt;chr&gt;, MICA &lt;chr&gt;,
#   MICB &lt;chr&gt;</code></pre>
</div>
</div>
<p>Gene copies can be separated into two columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This function needs to be uploaded.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other functions will deal with ambiguities either by keeping the first ambiguity and discarding the others, or placing the other ambiguities into a different column. These functions need to be uploaded.</p>
<p>The <code>HLA_truncate</code> function takes molecular typing and truncates it to a specified number of fields. This is useful for performing matching between alleles that may be at different resolution. Currently the function works only on a single allele, but we plan to update it to work on all alleles in a GL string.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="st">"A*02:01:02:03N"</span> <span class="sc">%&gt;%</span> <span class="fu">HLA_truncate</span>(<span class="at">fields =</span> <span class="dv">2</span>, <span class="at">keep_suffix =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A*02:01N"</code></pre>
</div>
</div>
</section>
<section id="matchingmismatching" class="level3">
<h3 class="anchored" data-anchor-id="matchingmismatching">Matching/Mismatching</h3>
<p>Performing matching, or determining mismatches, is a common need for HLA data. We are developing a set of functions to perform matching. Busra has started with the HvG functions:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(immunogenetr)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>recip_genotype <span class="ot">&lt;-</span> <span class="st">"HLA-A*01:01+HLA-A*02:01"</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>donor_genotype <span class="ot">&lt;-</span> <span class="st">"HLA-A*02:01+HLA-A*03:01"</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">HLA_mismatch_HvG</span>(recip_genotype, donor_genotype, <span class="st">"A"</span>) <span class="co"># The first function gives a logical output if a mismatch is present.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HLA_mismatch_number_HvG</span>(recip_genotype, donor_genotype, <span class="st">"A"</span>) <span class="co"># The `number` function gives an integer [0-2] as the output.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">HLA_mismatch_alleles_HvG</span>(recip_genotype, donor_genotype, <span class="st">"A"</span>) <span class="co"># The `alleles` function give the mismatched allele(s) as the output.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "HLA-A*03:01"</code></pre>
</div>
</div>
<p>Once we finalize how these mismatching functions work, we plan to write functions for GvH direction mismatches, as well as solid organ mismatching (which will just be an alias for HvG). We will also have to write matching functions. Matching functions will require both GvH, HvG and bidirectional matching. These functions can serve as the basis for high-level functions, for example a BMT matching function that would take “X/8” as an argument and perform matching at the HLA-A, B, C, and DRB1 loci.</p>
<p>One question I have is how to handle ambiguity in the matching and mismatching functions. Some examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>recip_genotype <span class="ot">&lt;-</span> <span class="st">"HLA-DPB1*03:01+HLA-DPB1*04:01"</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>donor_genotype <span class="ot">&lt;-</span> <span class="st">"HLA-DPB1*03:01+HLA-DPB1*04:01/HLA-DPB1*126:01"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I think it’s pretty clear that if we saw this result clinically, for example when reviewing a deceased donor typing report, we would conclude that there is not a mismatch. However, if the real donor typing is HLA-DPB1*03:01+HLA-DPB1*126:01, then it would be a mismatch. Except, if we perform matching at the G group level, then it would not be a mismatch. For this reason I think it is best to have these simple match/mismatch functions require an unambiguous GL string, and rely on the researcher to perform imputation or P group-level translation first, as any assumption we build into these functions may not be the correct.</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>