---
title: "Function tests"
---

```{r}
# These are the libraries needed for the functions.

library(dplyr)
library(tidyselect)
library(stringr)
library(tidyr)
library(xml2)
library(rlang)
library(cli)
library(purrr)
```

```{r}
# Open the test data
(test1 <- read.csv("test1.csv"))
(test2 <- read.csv("test2.csv"))
(test3 <- read.csv("test3.csv"))
(test4 <- read.csv("test4.csv"))
(test5 <- readRDS("test5.rds"))
```

```{r}
# Test the HLA_column_repair function

(test <- tribble(
  ~case, ~"HLA-A*", ~"HLA-B", ~"HLA-DRB1*", ~"HLA-DQB1",
  1, "A2", "A3", "A2", "A68",
  2, "2", "3", "3", "2",
  3, "01:01", NA, "01:01", "01:01",
  4, "01:02:06", NA, NA, "01:02:06"
))

test %>% HLA_column_repair()
test %>% HLA_column_repair(format = "tidyverse", asterisk = FALSE)
test %>% HLA_column_repair(format = "WHO", asterisk = TRUE)
test %>% HLA_column_repair(format = "R", asterisk = FALSE)
```

```{r}
# HLA genomic truncate function: truncate multi-field typing to fewer fields. If any number other than 1, 2, or 3 is entered for "fields," the entire type will be returned. Similarly, non-molecular nomenclature will be returned without modification. Note this will not keep G or P group designations, as they are not WHO-recognized suffixes. Should these also be kept? Maybe have a separate argument to keep or remove them?

HLA_truncate <- function(data, fields = 2, keep_suffix = TRUE) {
  # Split the input GL string into individual alleles
  alleles <- strsplit(data, "\\^")[[1]]
  
  # Initialize an empty vector to store truncated alleles
  truncated_alleles <- vector("character", length(alleles))
  
  # Loop through each allele to apply truncation
  for (i in seq_along(alleles)) {
    allele <- alleles[i]
    
    # Extract fields based on the number of fields specified
    A <- if (fields == 3) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}:?[:digit:]{0,4}")
    } else if (fields == 2) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}")
    } else if (fields == 1) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}")
    } else {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}:?[:digit:]{0,4}:?[:digit:]{0,4}")
    }
    
    # Extract any WHO-recognized suffixes
    B <- replace_na(str_extract(allele, "[LSCAQNlscaqn]$"), "")
    
    # Glue truncated typing and suffixes if indicated
    if (keep_suffix == TRUE) {
      truncated_alleles[i] <- na_if(str_c({A}, {B}, sep = ""), "NA")
    } else {
      truncated_alleles[i] <- A
    }
  }
  
  # Join the truncated alleles back into a single string
  truncated_string <- paste(truncated_alleles, collapse = "^")
  
  return(truncated_string)
}




# A function to take HLA typing data spread across different columns, as is often found in wild-caught data, and transform it to a GL string.

HLA_columns_to_GLstring <- function(data, case_column, HLA_typing_columns, prefix_to_remove = "", suffix_to_remove = "", serologic = FALSE){
  # Set up prefix and suffix regex
  prefix <- str_c("^", str_escape(prefix_to_remove))
  suffix <- str_c(str_escape(suffix_to_remove), "$")
  # Identify the columns to modify
  col2mod <- names(select(data, {{HLA_typing_columns}}))
  
  step1 <- data %>% 
    # pivoting longer to get each allele on a separate row.
    pivot_longer(cols = all_of(col2mod), names_to = "names", values_to = "allele") %>% 
    # Remove any prefixes or suffixes from locus names
    mutate(trunctated_names = str_replace(names, regex(prefix, ignore_case = TRUE), "")) %>% 
    mutate(trunctated_names = str_replace(trunctated_names, regex(suffix, ignore_case = TRUE), "")) %>% 
    # Use the HLA_truncate function to clean up the typing, and remove any blank values.
    mutate(allele = HLA_truncate(allele, fields = 4, keep_suffix = TRUE)) %>% 
    # Remove any blank (now NA) typing values.
    filter(!is.na(allele)) %>% 
    # Determine the locus from the column names
    mutate(locus_from_name = case_when(
      str_detect(trunctated_names, "[Dd]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[Dd][PpQqRr]?[AaBb]?[:digit:]?"))),  #If the column was named "DRB345", for example, then all of the locus names will be "DRB3." Will have to extract names from allele to determine which locus.
      str_detect(trunctated_names, regex("TAP", ignore_case = TRUE)) ~ str_to_upper(str_extract(trunctated_names, "TAP[12]")),
      str_detect(trunctated_names, regex("HFE", ignore_case = TRUE)) ~ "HFE",
      str_detect(trunctated_names, regex("MIC", ignore_case = TRUE)) ~ str_to_upper(str_extract(trunctated_names, "MIC[AB]")),
      str_detect(trunctated_names, "(?<![LlQqPpRrCcTt])[Aa]") ~ "HLA-A",
      str_detect(trunctated_names, "(?<![QqPpRrCc])[Bb]") ~ "HLA-B",
      str_detect(trunctated_names, "(?<![QqPpRrIi])[Cc]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[Cc][Ww]?"))),
      str_detect(trunctated_names, "[EFGHJKLNPSTUVWYefghjklnpstuvwy]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[EFGHJKLNPSTUVWYefghjklnpstuvwy]"))),
      .default = "unknown"
    )) %>% 
    # Rename any "CW" properly
    mutate(locus_from_name = str_replace(locus_from_name, "HLA-CW", "HLA-Cw")) %>% 
    # Determine the DRB locus from the allele
    mutate(DRB_locus = if_else(
      str_detect(locus_from_name, "DRB"), str_c("HLA-DRB", str_extract(allele, "[1345](?=\\*)")), NA_character_
    )) %>% 
    # Determine the final locus name from the two columns
    mutate(final_locus = coalesce(DRB_locus, locus_from_name)) %>% 
    # Remove any info before the asterisk from the allele field
    mutate(allele = str_replace(allele, ".+\\*", "")) 
  
  # Set up error detection for any loci that could not be determined
  error_table <- step1 %>% filter(locus_from_name == "unknown")
  error_column_names <- error_table %>% select(names) %>% distinct() %>% dplyr::pull(names)
  
  # Error code
  if (nrow(error_table) != 0) {
    abort(format_error("The  column(s) {error_column_names} could not be parsed to determine HLA loci."))
  }
  
  # Assemble the final type
  if (serologic == TRUE){
    step2 <- step1 %>% mutate(final_type = str_glue("{final_locus}{allele}", .na = ""))
  } else {
    step2 <- step1 %>% mutate(final_type = str_glue("{final_locus}*{allele}", .na = ""))
  }
  
  # Assemble the GL string
  step2 %>% summarise(final_type_2 = str_flatten(final_type, collapse = "+"), .by = c({{case_column}}, locus_from_name, DRB_locus)) %>% 
    summarise(GL_string = str_flatten(final_type_2, collapse = "^"), .by = c({{case_column}})) %>% dplyr::pull(GL_string)
}

typing <- "A*01:01:01:02N^B*07:02:01:01Q^C*04:01:01:01A"
HLA_truncate(typing, fields = 2, keep_suffix = TRUE) 
# Expected output: "A*01:01N^B*07:02Q^C*04:01A"
```

```{r}
# Test the HLA_columns_to_GLstring function

(test_a <- tribble(
  ~case, ~"a1", ~"a2", ~"drw1", ~"drw2", ~"cw1", ~"cw2",
  1, "1", "2", "51", "52", "1", "15",
  2, "2", "3", "53", "blank", "10", "blank",
  3, "2", NA, "51", "51", "3", NA,
  4, "68", NA, NA, "53", "5", "17"
))

(test_b <- tribble(
  ~case, ~"A1", ~"A2", ~"DRBX1", ~"DRBX2", ~"E1", ~"E2", ~"MICB1", ~"MICA2", ~"TAP11", ~"TAP22", ~"HFE1", ~"HFE2", ~"S1",
  1, "01:01", "02:01", "3*01:01", "5*02:01", "01:01", "02:01", "001:01", "002:02", "01:01", "03:01", "001:01", "002", "001:001",
  2, "02:02", "03:01", "4*02:01", "blank", "03:01", "blank", "001:02", "002:01", "01:03", "blank", "001:01", "missing", NA,
  3, "68:02", NA, "3*02:02", "4*03:01N", "04:01", NA, "001:01", "blank", "05:01", NA, "004", NA, "blank",
  4, "11:01", NA, NA, "4*03", "01:01", "05:01", "002:03", NA, "06:01", "01:01", "003", NA, "0101"
))

(test4 %>% mutate(GL_string = HLA_columns_to_GLstring(., patient, c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient")))
(test1 %>% mutate(GL_string = HLA_columns_to_GLstring(., patient, c(A1:DPB1_2)))
  %>% select(patient, DRB345_1:DRB345_2, GL_string)
)
(test_a %>% mutate(GL_string = HLA_columns_to_GLstring(., case, c(a1:cw2), serologic = TRUE)))
(test_b %>% mutate(GL_string = HLA_columns_to_GLstring(., case, c(A1:S1))))
```


```{r}
str_view("HLA-DRB3", "DRB")

#Test
#Test2
```

```{r}
# Test HLA_mismatch_HvG function
test_cases <- tribble(
  ~case, ~GL_string_recip, ~GL_string_donor, ~locus, ~expected_result,
  1, "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01", "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01", "DQB1", TRUE,
  2, "HLA-A*24:02+HLA-A*29:02", "HLA-A*02:01", "A", TRUE,
  3, "HLA-A*02:01", "HLA-A*02:01+HLA-A*11:05", "HLA-A", TRUE,
  4, "HLA-A*02:01+HLA-A*11:05", "HLA-A*02:01", "A", FALSE
)
test_cases <- test_cases %>%
  rowwise() %>%
  mutate(result = HLA_mismatch_HvG(GL_string_recip, GL_string_donor, locus),
         passed = result == expected_result)

print(test_cases)

GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"

print(HLA_mismatch_HvG(GL_string_recip, GL_string_donor, c("A", "DQB1", "HLA-DRB1")))

```

```{r}
# Test HLA_mismatch_alleles_HvG function
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*03:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
loci <- c("A", "DQB1", "HLA-DRB1")
#made DRB1 homozygous in donor to test homozygous_count parameter

print(HLA_mismatch_alleles_HvG(GL_string_recip, GL_string_donor, loci, homozygous_count = 1))

```

```{r}
# Test HLA_mismatch_number_HvG function
test_cases <- tribble(
  ~case, ~GL_string_recip, ~GL_string_donor, ~locus, ~expected_result,
  1, "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01", "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01", "HLA-DQB1", 2,
  2, "HLA-A*24:02+HLA-A*29:02", "HLA-A*02:01", "HLA-A", 1,
  3, "HLA-A*02:01", "HLA-A*02:01+HLA-A*11:05", "A", 1,
  4, "HLA-A*02:01+HLA-A*11:05", "HLA-A*02:01", "A", 0
)
test_cases <- test_cases %>%
  rowwise() %>%
  mutate(result = HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, locus),
         passed = result == expected_result)

print(test_cases)

GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*03:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
#made DRB1 homozygous in donor to test homozygous_count parameter

print(HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, c("A", "DQB1", "HLA-DRB1"), homozygous_count = 2))
```

```{r}
# Test GLstring_genes and GLstring_genes_expanded
GLstring_genes <- tibble(GL_string = "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01") %>%
  GLstring_genes("GL_string")

GLstring_genes_expanded <- tibble(GL_string = "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01") %>%
  GLstring_genes_expanded("GL_string")

print(GLstring_genes)
print(GLstring_genes_expanded)

```

```{r}
HLA_mismatch_number_HvG <- function(GL_string_recip, GL_string_donor, loci) {
  # Check for ambiguity
  if (str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]")) {
    stop("HLA_mismatch_number_HvG does not support ambiguous GL strings that contain the delimiters | or /")
  }
  
  # Normalize the loci input
  loci <- gsub("HLA_", "", loci)  # Remove HLA_ if present
  loci <- gsub("HLA-", "", loci)  # Remove HLA- if present
  
  # Process recipient and donor GL strings
  recip_data <- tibble(GL_string = GL_string_recip) %>%
    GLstring_genes_expanded("GL_string")
  donor_data <- tibble(GL_string = GL_string_donor) %>%
    GLstring_genes_expanded("GL_string")
  
  # Initialize a list to store mismatch counts for each locus
  mismatch_counts <- list()
  
  # Check mismatch for each locus
  for (locus in loci) {
    # Check if the specified locus exists in both datasets
    if (!(locus %in% names(recip_data)) | !(locus %in% names(donor_data))) {
      stop(paste("Locus", locus, "not found in both recipient and donor data."))
    }
    
    # Extract unique entries for the specified locus from recipient and donor data
    recip_locus_entries <- recip_data %>% pull({{ locus }}) %>% na.omit() %>% unique()
    donor_locus_entries <- donor_data %>% pull({{ locus }}) %>% na.omit() %>% unique()
    
    # Identify mismatched alleles
    mismatches <- donor_locus_entries[!donor_locus_entries %in% recip_locus_entries]
    
    # Store the number of mismatches
    mismatch_counts[[locus]] <- length(mismatches)
  }
  
  # Return result
  if (length(loci) == 1) {
    return(mismatch_counts[[loci]])
  } else {
    return(paste(sapply(names(mismatch_counts), function(locus) {
      paste0(locus, ":", mismatch_counts[[locus]])
    }), collapse = ","))
  }
}


GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01"
GL_string_donor <- "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
locus <- c("A", "B", "C", "DRB1")
HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, locus)

```



```{r}
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02"
GL_string_donor_homozygous <- "HLA-A*03:01+HLA-A*03:01"
GL_string_donor_heterozygous <- "HLA-A*03:01+HLA-A*68:01"

HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor_homozygous, "A")
HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor_heterozygous, "A")
```

```{r}
GL_string_molecular <- "HLA-A*01:01:01:01N/HLA-A*01:02x/HLA-A*01:03Q/HLA-A*01:95+HLA-A*24:02:01:01|HLA-A*01:01:01:01/HLA-A*01:03P+HLA-A*24:03:01:01^HLA-B*07:01:01G+HLA-B*15:01:01/HLA-B*15:02:01|HLA-B*07:03+HLA-B*15:99:01^HLA-DRB1*03:01:02~HLA-DRB5*01:01:01+HLA-KIR2DL5A*0010101+HLA-KIR2DL5A*0010201?HLA-KIR2DL5B*0010201+HLA-KIR2DL5B*0010301"

GL_string_sero <-"HLA-A2+HLA-A68^HLA-DR17+HLA-DR18"

HLA_truncate(GL_string_molecular, fields = 3, keep_suffix = TRUE, keep_G_P_group = FALSE)
HLA_truncate(GL_string_sero, fields = 3)

HLA_truncate("HLA-A*01:02v")
HLA_truncate("blank")
```


```{r}
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02"
GL_string_donor <- "HLA-A*03:01+HLA-A*30:01"

HLA_mismatch_HvG(GL_string_recip, GL_string_donor, c("A"))
```

```{r}
# HLA_validate function to return only HLA alleles in valid nomenclature, either serologic or molecular. Simple numbers, such as "2" or "27" will be returned as-is. Suffixes that are not WHO-recognized suffixes (L, S, C, A, Q, N) or G or P group designations will be removed. For example "novel" at the end of the allele will be removed, while "n" at the end of the allele will be retained. Other values, such as "blank" or "-" will be converted to NA values. This function is helpful for cleaning up the typing of an entire table of HLA values.

HLA_validate <- function(data){
  data %>% str_extract("(HLA-)?([:alnum:]*)(\\*)?[:digit:]{1,}:?[:digit:]*:?[:digit:]*:?[:digit:]*([GPLSCAQNgplscaqn](?!.))*") %>% as.character()
}

HLA_validate("HLA-A2")
HLA_validate("A*02:01:01:01N")
HLA_validate("A*02:01:01N")
HLA_validate("HLA-DRB1*02:03novel")
HLA_validate("HLA-DQB1*03:01v")
HLA_validate("HLA-DRB1*02:03P")
HLA_validate("HLA-DPB1*04:01:01G")
HLA_validate("2")
HLA_validate(2)
HLA_validate("B27")
HLA_validate("A*010101")
HLA_validate("-")
HLA_validate("blank")
```

```{r}
# read_HML function

read_HML <- function(HML_file){
  # Validate input
  if (!file.exists(HML_file)) {
    stop("The file does not exist:", HML_file)
  }  
  
  # Load the HML file 
  HML <- tryCatch({
    read_xml(HML_file)
  }, error = function(e){
    stop("Failed to read HML; check that file is in compliant HML format.")
  })
  
  # Filter for all the children in the HML file that represent a sample
  samples <- xml_find_all(HML, ".//d1:sample")
  
  # Get sample number and GL strings for each sample
  GL_strings <- map(samples, function(node){
    # Get sample ID 
    sampleID <-  xml_attr(node, "id")
    # Get GL strings 
    glstring <- xml_text(xml_find_all(node, ".//d1:glstring"))  
    # Combine to a tibble
    tibble(sampleID, glstring)
  }) 
  
  # Combine to a single tibble.
  combined <- bind_rows(GL_strings)
  
  # Some implementations of HML put the same locus in multiple nodes; this combines them with "+" to form a compliant GL string
  reduced <- combined %>% 
    mutate(locus = str_extract(glstring, "[^//*]+")) %>% 
    mutate(glstring = paste0(glstring, collapse = "+"), .by = c(sampleID, locus)) %>% 
    # Clean up values
    distinct(sampleID, glstring, locus) %>% 
    filter(!is.na(sampleID) & !is.na(glstring)) %>% 
    select(-locus) 
  
  # Combine to a single GL string per sample
  summarise(reduced, GL_string = str_flatten(glstring, collapse = "^"), .by = sampleID)
  
  
  
}

read_HML("HML_1.hml")
read_HML("HML_2.hml")
read_HML("test1.csv")
```

```{r}
# Troubleshoot HLA_truncate error

HLA_truncate <- function(data, fields = 2, keep_suffix = TRUE, keep_G_P_group = FALSE) {
  # Expand the GL string
  alleles <- GLstring_expand_longer(data) %>%
    # Extract any WHO-recognized suffixes
    mutate(suffix = replace_na(str_extract(value, "(?<=[:digit:])[LSCAQNlscaqn]$"), "")) %>%
    # Extract any P or G group designation
    mutate(GP = replace_na(str_extract(value, "(?<=[:digit:])[PGpg]$"), "")) %>%
    # Separate HLA prefix if available
    separate_wider_delim(value, delim = "-", names = c("prefix", "rest"), too_few = "align_end") %>% 
    separate_wider_delim(rest, delim = "*", names = c("gene", "molecular_type"), too_few = "align_start") %>% 
    mutate(sero_type = str_extract(gene, "[:digit:]+$"), .after = molecular_type) %>% 
    mutate(gene = str_replace(gene, "[:digit:]+$", "")) %>%
    # Separate molecular fields
    separate_wider_delim(molecular_type, delim = ":", names = c("one", "two", "three", "four"), too_few = "align_start") %>%
    # Keep only numbers in each field, in case there were non-standard suffixes.
    mutate(across(one:four, ~str_extract(., "[:digit:]+")))
  
  # Delete fields for truncating and reunite the alleles
  if (fields == 1) {
    trunctated <- alleles %>% select(-four, -three, -two) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(gene, gene, one, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
  } else if (fields == 2) {
    trunctated <- alleles %>% select(-four, -three) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:two, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
  } else if (fields == 3) {
    trunctated <- alleles %>% select(-four) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:three, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
  } else {
    trunctated <- alleles %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:four, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
  }
  
  # Retain suffix if desired
  if (keep_suffix) {
    with_suffix <- trunctated %>% unite(gene, gene, suffix, sep = "", na.rm = TRUE)
  } else {
    with_suffix <- trunctated %>% select(-suffix)
  }
  # Retain P/G group designation if desired
  if (keep_G_P_group) {
    with_g_p <- with_suffix %>% unite(gene, gene, GP, sep = "", na.rm = TRUE)
  } else {
    with_g_p <- with_suffix %>% select(-GP)
  }
  
  # Combine everything back to a GL string.
  final <- with_g_p %>% rename(value = gene) %>% ambiguity_table_to_GLstring()
  return(final)
}

HLA_truncate(c("A2", "HLA-A*68:01:02N", "DQA01", "DP4", "DR51"), fields = 2)
```

```{r}
GLstring_expand_longer <- function(GL_string){
  as_tibble(GL_string) %>%
    # Assign a unique identifier for each entry for the function
    mutate(entry = row_number()) %>%
    # Separate GL string precedence 0: possible gene locations
    separate_longer_delim(value, delim = "?") %>%
    # Assign a unique identifier for each possible gene location
    mutate(possible_gene_location = row_number(), .by = entry) %>%
    # Separate GL string precedence 1: gene/locus
    separate_longer_delim(value, delim = "^") %>%
    # Identify locus
    mutate(locus = row_number(), .by = c(entry, possible_gene_location)) %>%
    # Separate GL string precedence 2: genotype list
    separate_longer_delim(value, delim = "|") %>%
    # Identify genotype ambiguities
    mutate(genotype_ambiguity = row_number(), .by = c(entry, possible_gene_location, locus)) %>%
    # Separate GL string precedence 3: genotype
    separate_longer_delim(value, delim = "+") %>%
    # Identify genotypes
    mutate(genotype = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity)) %>%
    # Separate GL string precedence 4: haplotype
    separate_longer_delim(value, delim = "~") %>%
    # Identify haplotypes
    mutate(haplotype = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity, genotype)) %>%
    # Separate GL string precedence 5: allele list
    separate_longer_delim(value, delim = "/") %>%
    # Identify alleles
    mutate(allele = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity, genotype, haplotype))
}

GLstring_expand_longer(c("A*02:01", "A2"))
```

```{r}
# Work on the HLA_columns_to_GLstring function some more.

HLA_columns_to_GLstring <- function(data, HLA_typing_columns, prefix_to_remove = "", suffix_to_remove = "") {
  # Set up prefix and suffix regex to remove unwanted parts
  prefix_regex <- regex(str_c("^", str_escape(prefix_to_remove)), ignore_case = TRUE)
  suffix_regex <- regex(str_c(str_escape(suffix_to_remove), "$"), ignore_case = TRUE)
  
  # Identify the columns to modify
  col2mod <- names(select(data, {{HLA_typing_columns}}))
  
  # Processing Step 1
  step1 <- data %>%
    mutate(row_for_function = 1:nrow(.)) %>%
    # pivoting longer to get each allele on a separate row.
    pivot_longer(cols = all_of(col2mod), names_to = "names", values_to = "allele") %>% 
    # Determine if typing is serologic by presence of ":" in allele.
    mutate(molecular = str_detect(allele, ":") | str_detect(allele, "^0") | str_detect(names, "(DQA1)|(DPB1)|(DPA1)")) %>%
    # Remove prefixes and suffixes from column names
    mutate(truncated_names = str_replace(names, prefix_regex, ""),
           truncated_names = str_replace(truncated_names, suffix_regex, "")) %>%
    # Use the HLA_validate function to clean up the typing
    mutate(allele = HLA_validate(allele)) %>%
    # Remove any prefixes in alleles
    HLA_prefix_remove(allele) %>%
    # Determine locus name using extended logic
    mutate(locus_from_name = case_when(
      str_detect(truncated_names, regex("^A[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-A",  # Handle A locus
      str_detect(truncated_names, regex("^Bw[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-Bw",  # Handle Bw separately from B locus
      str_detect(truncated_names, regex("^B[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-B",  # Handle B locus
      str_detect(truncated_names, regex("^Cw[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-Cw",  # Handle Cw locus
      str_detect(truncated_names, regex("^C[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-C",  # Handle C locus
      str_detect(truncated_names, regex("^DRB1[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DRB1",  # Handle molecular DRB1 locus
      str_detect(truncated_names, regex("^DRB3[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DRB3",  # Handle molecular DRB3 locus
      str_detect(truncated_names, regex("^DRB4[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DRB4",  # Handle molecular DRB4 locus
      str_detect(truncated_names, regex("^DRB5[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DRB5",  # Handle molecular DRB5 locus
      str_detect(truncated_names, regex("^DRw?[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DR",  # Handle serologic DR locus
      str_detect(truncated_names, regex("^DQB1[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DQB1",  # Handle molecular DQB1 locus
      str_detect(truncated_names, regex("^DQA1[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DQA1",  # Molecular DQA1
      str_detect(truncated_names, regex("^DQ[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DQ",  # Handle serologic DQ locus
      str_detect(truncated_names, regex("^DPA1[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DPA1",  # Molecular DPA1
      str_detect(truncated_names, regex("^DPB1[:digit:]?\\*?", ignore_case = TRUE)) ~ "HLA-DPB1",  # Molecular DPB1
      TRUE ~ "unknown"
    )) %>%
    # Determine serologic locus names
    mutate(serologic_name = case_when(
      locus_from_name == "HLA-A" ~ "HLA-A",
      locus_from_name == "HLA-B" ~ "HLA-B",
      locus_from_name == "HLA-Bw" ~ "HLA-Bw",
      str_detect(locus_from_name, "HLA-C[Ww]?") ~ "HLA-Cw",
      str_detect(locus_from_name, "DR") ~ "HLA-DR",
      locus_from_name == "HLA-DQA1" ~ "HLA-DQA",
      str_detect(locus_from_name, "HLA-DQ") ~ "HLA-DQ",
      locus_from_name == "HLA-DPA1" ~ "HLA-DPA",
      locus_from_name == "HLA-DPB1" ~ "HLA-DP",
      TRUE ~ "unknown"
    )) %>% 
    # Determine if the DR typing is one of the DR51/52/53 loci
    mutate(DRB345 = if_else(str_detect(allele, "^5") & locus_from_name == "HLA-DR", TRUE, FALSE)) %>%
    # Handle the final locus name for high-resolution DRB alleles
    mutate(DRB_locus = if_else(
      str_detect(locus_from_name, "DRB"), str_c("HLA-DRB", str_extract(allele, "[1345](?=\\*)")), NA_character_
    )) %>%
    # Determine the final molecular locus name
    mutate(molecular_locus = coalesce(DRB_locus, locus_from_name)) %>%
    # Record "XX" if there is no typing at any of the selected loci.
    mutate(allele = if(all(is.na(allele))) "XX" else allele, .by = c(row_for_function, molecular_locus)) %>%
    # Remove any blank values
    filter(!is.na(allele)) %>%
    # Remove any "XX" values from the DRB3/4/5 loci
    filter(!(allele == "XX" & str_detect(molecular_locus, "DRB[345]")))

  # Set up error detection for any loci that could not be determined
  error_table <- step1 %>% filter(locus_from_name == "unknown")
  error_column_names <- error_table %>% select(names) %>% distinct() %>% dplyr::pull(names)

  if (nrow(error_table) != 0) {
    # Print the columns that caused the error to assist in debugging
    print(glue::glue("Columns that could not be parsed: {paste(error_column_names, collapse = ', ')}"))
    abort(format_error("The column(s) {paste(error_column_names, collapse = ', ')} could not be parsed to determine HLA loci."))
  }

  # Assemble the final type
  step2 <- step1 %>%
    mutate(final_type = if_else(
      molecular,
      str_glue("{molecular_locus}*{allele}"),
      str_glue("{serologic_name}{allele}")
    )) %>%
    # Group alleles within the same locus
    summarise(final_type_2 = str_flatten(final_type, collapse = "+"), .by = c(row_for_function, locus_from_name, DRB345)) %>%
    # Assemble GL string with each locus separated by "^"
    summarise(GL_string = str_flatten(final_type_2, collapse = "^"), .by = row_for_function)

  # Return GL Strings
  return(step2 %>% dplyr::pull(GL_string))
    
}

typing_example <- readRDS("typing_example.rds")

(test5 %>% mutate(GL_string = HLA_columns_to_GLstring(., c(A1Cd:mDPB12cd), prefix_to_remove = "m", suffix_to_remove = "Cd")))
(test4 %>% mutate(GL_string = HLA_columns_to_GLstring(., c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient")))
(test3 %>% mutate(GL_string = HLA_columns_to_GLstring(., c(HLA.A:HLA.DPB1), prefix_to_remove = "HLA.")))


(typing_example %>% mutate(GL_string = HLA_columns_to_GLstring(., c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient")))
# (typing_example %>% HLA_columns_to_GLstring(c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient"))
# (test5 %>% HLA_columns_to_GLstring(c(A1Cd:mDPB12cd), prefix_to_remove = "m", suffix_to_remove = "Cd"))
# (test4 %>% HLA_columns_to_GLstring(c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient"))
# (test3 %>% HLA_columns_to_GLstring(c(HLA.A:HLA.DPB1), prefix_to_remove = "HLA."))
```

# Below is a working mismatch function, which will serve as the basis for all other mismatching and matching functions. 

```{r}
HLA_mismatch_base <- function(GL_string_recip, GL_string_donor, loci, direction = c("HvG", "GvH"), homozygous_count = 2) {
  # Ensure input vectors are of the same length - each input should be a single GL string.
  if (length(GL_string_recip) != length(GL_string_donor)) {
    stop("Recipient and donor GL strings must be of equal length")
  }
  
  # Check for ambiguity
  if (any(str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]") )) {
    stop("The matching/mismatching functions do not support ambiguous GL strings containing | or /. Process your GL strings to result in unambiguous genotypes before using these functions.")
  }
  
  # Function to preprocess GL strings: handle null alleles and homozygosity
  preprocess_GL_string <- function(GL_string, homozygous_count) {
    # Split GL string into alleles
    alleles <- str_split(GL_string, "\\+", simplify = TRUE)
    
    # Replace any allele that ends with "N" with "NullN".
    alleles <- str_replace(alleles, "(?<=HLA-[:alnum:]{1,4}\\*).+N$", "XXN")
    
    # Handle homozygosity
    if (length(alleles) == 1 && homozygous_count == 2) {
      alleles <- rep(alleles, times = homozygous_count)
    }
    
    # Return processed alleles as a single GL string
    return(str_flatten(alleles, collapse = "+", na.rm = TRUE))
    
  }
  
  # Function to process each pair of recipient/donor alleles
  process_pair <- function(recip_str, donor_str) {
    # Split GL strings by "^" to separate loci
    recip_alleles_list <- unlist(strsplit(recip_str, "\\^"))
    donor_alleles_list <- unlist(strsplit(donor_str, "\\^"))
    
    # Function to process the alleles strings.
    process_alleles <- function(alleles_list, homozygous_count) {
      # Process DRB345 alleles to add them to a single locus "HLA-DRB3/4/5"
      alleles_list_DRB345 <- alleles_list %>%
        keep(str_detect(., "HLA-DRB[345]")) %>%
        paste(collapse = "+")
      
      # Remove DRB345 alleles from the original list
      alleles_list_no_DRB345 <- alleles_list %>% 
        keep(!str_detect(., "HLA-DRB[345]"))
      
      # Combine lists and preprocess each allele string
      alleles_list_processed <- c(alleles_list_no_DRB345, alleles_list_DRB345) %>%
        keep(~ str_detect(.x, "[:graph:]")) %>%
        map_chr(preprocess_GL_string, homozygous_count)
      
      # Filter out any NA values
      alleles_list_processed <- alleles_list_processed[!is.na(alleles_list_processed)]
      
      return(alleles_list_processed)
    }
    
    # Process recipient and donor allele strings:
    recip_alleles_list_processed <- process_alleles(recip_alleles_list, homozygous_count)
    donor_alleles_list_processed <- process_alleles(donor_alleles_list, homozygous_count)
    
    # Set names for each locus
    names(recip_alleles_list_processed) <- map_chr(
      recip_alleles_list_processed,
      ~ if_else(str_detect(.x, "\\*"), # Molecular nomenclature
                strsplit(.x, "\\*")[[1]][1],
                str_extract(.x, "^HLA-[A-Za-z]+")) # Serologic nomenclature
    )
    names(donor_alleles_list_processed) <- map_chr(
      donor_alleles_list_processed,
      ~ if_else(str_detect(.x, "\\*"), # Molecular nomenclature
                strsplit(.x, "\\*")[[1]][1],
                str_extract(.x, "^HLA-[A-Za-z]+")) # Serologic nomenclature
    )
    
    # Modify the names if they start with DRB3, 4, 5
    names(recip_alleles_list_processed) <- modify_if(names(recip_alleles_list_processed), ~ str_detect(.x, "HLA-DRB[345]"), ~ "HLA-DRB3/4/5")
    names(donor_alleles_list_processed) <- modify_if(names(donor_alleles_list_processed), ~ str_detect(.x, "HLA-DRB[345]"), ~ "HLA-DRB3/4/5")
    
    # Modify the names for DR51, DR52, DR53 to treat them as a single locus
    names(recip_alleles_list_processed) <- modify_if(names(recip_alleles_list_processed), ~ str_detect(.x, "HLA-DR5[123]"), ~ "HLA-DR51/52/53")
    names(donor_alleles_list_processed) <- modify_if(names(donor_alleles_list_processed), ~ str_detect(.x, "HLA-DR5[123]"), ~ "HLA-DR51/52/53")
    
    # Aggregate DR51/52/53 alleles under the single locus "HLA-DR51/52/53"
    if ("HLA-DR51/52/53" %in% loci) {
      recip_alleles_list_processed["HLA-DR51/52/53"] <- recip_alleles_list %>%
        keep(str_detect(., "HLA-DR5[123]")) %>%
        paste(collapse = "+")
      donor_alleles_list_processed["HLA-DR51/52/53"] <- donor_alleles_list %>%
        keep(str_detect(., "HLA-DR5[123]")) %>%
        paste(collapse = "+")
    }
    
    # Find which supplied loci are missing from recipient and donor genotypes
    missing_loci_from_recipient <- setdiff(loci, names(recip_alleles_list_processed))
    missing_loci_from_donor <- setdiff(loci, names(donor_alleles_list_processed))

    if (length(missing_loci_from_recipient) > 0 || length(missing_loci_from_donor) > 0) {
      stop(paste("Either the recipient and/or donor GL strings are missing these loci:",
                 paste(union(missing_loci_from_recipient, missing_loci_from_donor), collapse = ", ")))
    }
    
    # Mismatch results calculation
    mismatch_results <- map(loci, function(locus_name) {
      # Pull out the allele list for each locus.
      recip_alleles_str <- recip_alleles_list_processed[locus_name]
      donor_alleles_str <- donor_alleles_list_processed[locus_name]
      
      if (direction == "GvH") {
        # Calculate matches (Including nulls)
        matched_allele_1 <- intersect(unlist(strsplit(donor_alleles_str,"\\+"))[1], unlist(strsplit(recip_alleles_str,"\\+")))
        matched_allele_2 <- intersect(unlist(strsplit(donor_alleles_str,"\\+"))[2], unlist(strsplit(recip_alleles_str,"\\+")))
        matched_alleles <- discard(c(matched_allele_1, matched_allele_2), is.na)
        # Calculate mismatches (excluding nulls)
        recip_valid <- unlist(strsplit(recip_alleles_str, "\\+"))
        recip_valid <- recip_valid[!str_detect(recip_valid, "[Nn]$")]
        mismatched_allele_1 <- setdiff(recip_valid[1], unlist(strsplit(donor_alleles_str,"\\+")))
        mismatched_allele_2 <- setdiff(recip_valid[2], unlist(strsplit(donor_alleles_str,"\\+")))
        mismatched_alleles <- discard(c(mismatched_allele_1, mismatched_allele_2), is.na)
        # Count number of matches and mismatches
        total_match_mismatch <- length(matched_alleles) + length(mismatched_alleles)
        # If total matches + mismatches < 2, and homozygous_count == 2, repeat mismatched alleles.
        if (total_match_mismatch < 2 && homozygous_count == 2) {
          mismatched_alleles <- rep(mismatched_alleles, times = homozygous_count)
         }
      } else if (direction == "HvG") {
        # Calculate matches (Including nulls)
        matched_allele_1 <- intersect(unlist(strsplit(recip_alleles_str,"\\+"))[1], unlist(strsplit(donor_alleles_str,"\\+")))
        matched_allele_2 <- intersect(unlist(strsplit(recip_alleles_str,"\\+"))[2], unlist(strsplit(donor_alleles_str,"\\+")))
        matched_alleles <- discard(c(matched_allele_1, matched_allele_2), is.na)
        # Calculate mismatches (excluding nulls)
        donor_valid <- unlist(strsplit(donor_alleles_str, "\\+"))
        donor_valid <- donor_valid[!str_detect(donor_valid, "[Nn]$")]
        mismatched_allele_1 <- setdiff(donor_valid[1], unlist(strsplit(recip_alleles_str,"\\+")))
        mismatched_allele_2 <- setdiff(donor_valid[2], unlist(strsplit(recip_alleles_str,"\\+")))
        mismatched_alleles <- discard(c(mismatched_allele_1, mismatched_allele_2), is.na)
        # Count number of matches and mismatches
        total_match_mismatch <- length(matched_alleles) + length(mismatched_alleles)
        # If total matches + mismatches < 2, and homozygous_count == 2, repeat mismatched alleles.
        if (total_match_mismatch < 2 && homozygous_count == 2) {
          mismatched_alleles <- rep(mismatched_alleles, times = homozygous_count)
         }
        } else {
        stop("Direction must either be 'GvH', or 'HvG'.")
      }
      
      # Create a string of mismatched alleles or 'NA' if no mismatches are found.
      allele_mismatches_str <-
        if (length(mismatched_alleles) > 0) {
          paste0(locus_name, "=", paste(mismatched_alleles, collapse = "+"))
        } else {
          paste0(locus_name, "=", "NA")
        }
    })
    
    result <- paste(mismatch_results, collapse = ", ")
    
    # If only a single locus was selected in the arguments, output without starting with the locus name followed by an equal sign.
    if (length(loci) == 1) {
      result %>% str_replace(str_c(loci, "="), "") %>% na_if("NA")
    } else {
      return(result)
    }
  }
  
  # Return final result by applying the GL strings to the process_pair function defined above.
  map2_chr(GL_string_recip, GL_string_donor, process_pair)
}


# This works as intended:
HLA_mismatch_base(
  "HLA-A*03:01+HLA-A*68:01", 
  "HLA-A*03:99N+HLA-A*03:08",
  c("HLA-A"),
  direction = "HvG", 
  homozygous_count = 2
)

HLA_mismatch_base(
  "HLA-A*03:01+HLA-A*74:01^HLA-B*53:01+HLA-B*57:03^HLA-C*04:01+HLA-C*07:01^HLA-DRB1*11:01+HLA-DRB1*16:02^HLA-DRB3*03:01^HLA-DRB5*02:21^HLA-DQA1*01:02+HLA-DQA1*05:05^HLA-DQB1*03:19+HLA-DQB1*05:02^HLA-DPA1*02:01+HLA-DPA1*02:02^HLA-DPB1*01:01", 
  "HLA-A*03:01+HLA-A*30:02^HLA-B*14:02^HLA-C*08:02^HLA-DRB1*01:02+HLA-DRB1*13:02^HLA-DRB3*03:01^HLA-DQA1*01:01+HLA-DQA1*01:02^HLA-DQB1*05:01+HLA-DQB1*06:09^HLA-DPA1*02:01^HLA-DPB1*05:01+HLA-DPB1*17:01",
  c("HLA-A", "HLA-DRB1", "HLA-DRB3/4/5"),
  direction = "GvH"
)

HLA_mismatch_base(
  "HLA-A2+HLA-A68^HLA-Cw1+HLA-Cw17^HLA-DR1+HLA-DR17^HLA-DR52^HLA-DPB1*04:01", 
  "HLA-A3+HLA-A69^HLA-Cw10+HLA-Cw9^HLA-DR4+HLA-DR17^HLA-DR52+HLA-DR53^HLA-DPB1*04:01+HLA-DPB1*04:02",
  # function works when "HLA-DR51/52/53" is removed
  c("HLA-A", "HLA-Cw","HLA-DR51/52/53", "HLA-DPB1"),
  direction = "HvG"
)

HLA_mismatch_base(
  "HLA-A*03:01+HLA-A*68:01", 
  "HLA-A*03:99N",
  c("HLA-A"),
  direction = "HvG", 
  homozygous_count = 2
)

(test_cases <- tribble(
  ~case, ~recip_type, ~donor_type,
  1,   "HLA-A*02:01+HLA-A*03:04^HLA-B*15:01^HLA-DRB3*01:01", "HLA-A*02:01+HLA-A*03:04^HLA-B*15:02",
  2,   "HLA-A*02:01+HLA-A*30:02^HLA-B*15:01+HLA-B*14:03", "HLA-A*02:17+HLA-A*32:01^HLA-B*15:01+HLA-B*14:03",
  3,   "HLA-A*11:01+HLA-A*23:01^HLA-B*15:01+HLA-B*14:03", "HLA-A*03:01+HLA-A*68:02^HLA-B*15:01+HLA-B*14:03"
))

#test_cases %>% mutate(mismatch = HLA_mismatch_base(recip_type, donor_type, c("HLA-A"), direction = "HvG", homozygous_count = 2))

HLA_mismatch_base(
  "HLA-A*03:01N+HLA-A*74:01", 
  "HLA-A*03:02N+HLA-A*30:02",
  c("HLA-A"),
  direction = "GvH"
)
```




```{r}
# Work on creating derivative mismatch functions
HLA_mismatch_number <- function(GL_string_recip, GL_string_donor, loci, direction = c("HvG", "GvH", "bidirectional"), homozygous_count = 2) {
  direction <- match.arg(direction, c("HvG", "GvH", "bidirectional"))
  
  # Compute mismatch strings for both HvG and GvH
  mismatch_HvG <- HLA_mismatch_base(GL_string_recip, GL_string_donor, loci, "HvG", homozygous_count)
  mismatch_GvH <- HLA_mismatch_base(GL_string_recip, GL_string_donor, loci, "GvH", homozygous_count)
  
  # Helper function to count mismatches per locus
  count_mismatches <- function(mismatch_string) {
    if (is.na(mismatch_string) || mismatch_string == "NA") {
      return(0L)
    }
    alleles <- unlist(strsplit(mismatch_string, "\\+"))
    return(as.integer(length(alleles)))
  }
  
  # Calculate mismatch counts for each locus
  calculate_mismatch_counts <- function(mismatch_result, loci) {
    if (length(loci) == 1) {
      map_int(mismatch_result, count_mismatches)
      } else {
      # Split the result by loci and count mismatches for each
      locus_mismatches <- strsplit(mismatch_result, ", ")[[1]]
      counts <- sapply(locus_mismatches, function(x) {
        locus_parts <- strsplit(x, "=")[[1]]
        if (length(locus_parts) == 2) {
          count_mismatches(locus_parts[2])
        } else {
          0
        }
      })
      names(counts) <- loci
      # Format the output for multiple loci
      return(paste(paste0(names(counts), "=", counts), collapse = ", "))
    }
  }
  
  # Process mismatch counts for both directions
  mismatch_count_HvG <- calculate_mismatch_counts(mismatch_HvG, loci)
  mismatch_count_GvH <- calculate_mismatch_counts(mismatch_GvH, loci)
  
  # Handle directions
  if (direction == "HvG") {
    return(mismatch_count_HvG)
  } else if (direction == "GvH") {
    return(mismatch_count_GvH)
  } else if (direction == "bidirectional") {
    if (length(loci) == 1) {
      return(as.integer(max(mismatch_count_HvG, mismatch_count_GvH, na.rm = TRUE)))
    } else {
      HvG_counts <- sapply(strsplit(mismatch_count_HvG, ", ")[[1]], function(x) as.integer(strsplit(x, "=")[[1]][2]))
      GvH_counts <- sapply(strsplit(mismatch_count_GvH, ", ")[[1]], function(x) as.integer(strsplit(x, "=")[[1]][2]))
      max_counts <- pmax(HvG_counts, GvH_counts, na.rm = TRUE)
      return(paste(paste0(loci, "=", max_counts), collapse = ", "))
    }
  }
  
  stop("Invalid direction") # This should not be reached due to match.arg
}




# This works properly:
HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01^HLA-DRB5*02:21", 
  "HLA-A*03:02+HLA-A*74:01^HLA-DRB3*03:01",
  c("HLA-A", "HLA-DRB3/4/5"),
  direction = "bidirectional"
)

# And so does this:
HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01^HLA-DRB5*02:21", 
  "HLA-A*03:02+HLA-A*20:01^HLA-DRB3*03:01",
  c("HLA-A", "HLA-DRB3/4/5"),
  direction = "bidirectional"
)

# This looks good:
HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01^HLA-DRB5*02:21", 
  "HLA-A*03:02+HLA-A*20:01^HLA-DRB3*03:01",
  c("HLA-A", "HLA-DRB3/4/5"),
  direction = "GvH"
)

# test bidirectional 
HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01+HLA-DRB3*04:01",
  "HLA-A*03:02^HLA-DRB3*03:01",
  c("HLA-A", "HLA-DRB3/4/5"),
  direction = "bidirectional"
)

HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01^HLA-DRB5*02:21", 
  "HLA-A*03:02+HLA-A*20:01^HLA-DRB3*03:01",
  c("HLA-A"),
  direction = "GvH"
)

HLA_mismatch_number(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01+HLA-DRB3*04:01",
  "HLA-A*03:02^HLA-DRB3*03:01",
  "HLA-DRB3/4/5",
  direction = "bidirectional"
)

# This works:
test_cases %>% mutate(mismatch = HLA_mismatch_number(recip_type, donor_type, c("HLA-A", "HLA-B"), direction = "bidirectional", homozygous_count = 2))

# And now this works, too.:
test_cases %>% mutate(mismatch = HLA_mismatch_number(recip_type, donor_type, c("HLA-A"), direction = "bidirectional", homozygous_count = 2))

```


```{r}

HLA_mismatch_logical <- function(GL_string_recip, GL_string_donor, loci, direction = c("HvG", "GvH", "bidirectional"), homozygous_count = 2) {
  direction <- match.arg(direction, c("HvG", "GvH", "bidirectional"))

  # Helper function to determine logical results
  mismatch_to_logical <- function(mismatch_number) {
    if (length(loci) == 1) {
      return(mismatch_number != 0)
    } else {
      # Convert mismatch counts to logicals for multiple loci
      locus_results <- strsplit(mismatch_number, ", ")[[1]]
      logical_results <- sapply(locus_results, function(x) {
        locus_parts <- strsplit(x, "=")[[1]]
        if (length(locus_parts) == 2) {
          locus_name <- locus_parts[1]
          mismatch_count <- as.integer(locus_parts[2])
          return(paste0(locus_name, "=", mismatch_count != 0))
        } else {
          return(paste0(x, "=FALSE"))
        }
      })
      return(paste(logical_results, collapse = ", "))
    }
  }

  # Get mismatch counts using HLA_mismatch_number
  mismatch_counts <- HLA_mismatch_number(GL_string_recip, GL_string_donor, loci, direction, homozygous_count)

  # Convert mismatch counts to logical results
  logical_results <- mismatch_to_logical(mismatch_counts)

  return(logical_results)
}

HLA_mismatch_logical(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01^HLA-DRB5*02:21", 
  "HLA-A*03:01^HLA-DRB3*03:01",
  c("HLA-A"),
  direction = "HvG"
)

HLA_mismatch_logical(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01+HLA-DRB3*04:01",
  "HLA-A*03:02^HLA-DRB3*03:01",
  "HLA-DRB3/4/5",
  direction = "bidirectional"
)

HLA_mismatch_logical(
  "HLA-A*03:01+HLA-A*74:01^HLA-DRB3*03:01+HLA-DRB3*04:01",
  "HLA-A*03:02^HLA-DRB3*03:01",
  c("HLA-A", "HLA-DRB3/4/5"),
  direction = "bidirectional"
)

```



