---
title: "Function tests"
output: html_notebook
---

```{r}
# These are the libraries needed for the functions.

library(dplyr)
library(tidyselect)
library(stringr)
library(tidyr)
library(xml2)
library(rlang)
library(cli)
library(purrr)
```

```{r}
# Open the test data
(test1 <- read.csv("test1.csv"))
(test2 <- read.csv("test2.csv"))
(test3 <- read.csv("test3.csv"))
(test4 <- read.csv("test4.csv"))
```

```{r}
# Test the HLA_column_repair function

(test <- tribble(
  ~case, ~"HLA-A*", ~"HLA-B", ~"HLA-DRB1*", ~"HLA-DQB1",
  1, "A2", "A3", "A2", "A68",
  2, "2", "3", "3", "2",
  3, "01:01", NA, "01:01", "01:01",
  4, "01:02:06", NA, NA, "01:02:06"
  ))

test %>% HLA_column_repair()
test %>% HLA_column_repair(format = "tidyverse", asterisk = FALSE)
test %>% HLA_column_repair(format = "WHO", asterisk = TRUE)
test %>% HLA_column_repair(format = "R", asterisk = FALSE)
```

```{r}
# HLA genomic truncate function: truncate multi-field typing to fewer fields. If any number other than 1, 2, or 3 is entered for "fields," the entire type will be returned. Similarly, non-molecular nomenclature will be returned without modification. Note this will not keep G or P group designations, as they are not WHO-recognized suffixes. Should these also be kept? Maybe have a separate argument to keep or remove them?

HLA_truncate <- function(data, fields = 2, keep_suffix = TRUE) {
  # Split the input GL string into individual alleles
  alleles <- strsplit(data, "\\^")[[1]]

  # Initialize an empty vector to store truncated alleles
  truncated_alleles <- vector("character", length(alleles))

  # Loop through each allele to apply truncation
  for (i in seq_along(alleles)) {
    allele <- alleles[i]

    # Extract fields based on the number of fields specified
    A <- if (fields == 3) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}:?[:digit:]{0,4}")
    } else if (fields == 2) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}")
    } else if (fields == 1) {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}")
    } else {
      str_extract(allele, "(HLA-)?([:alnum:]{0,4})(\\*)?[:digit:]{1,4}:?[:digit:]{0,4}:?[:digit:]{0,4}:?[:digit:]{0,4}")
    }

    # Extract any WHO-recognized suffixes
    B <- replace_na(str_extract(allele, "[LSCAQNlscaqn]$"), "")

    # Glue truncated typing and suffixes if indicated
    if (keep_suffix == TRUE) {
      truncated_alleles[i] <- na_if(str_c({A}, {B}, sep = ""), "NA")
    } else {
      truncated_alleles[i] <- A
    }
  }

  # Join the truncated alleles back into a single string
  truncated_string <- paste(truncated_alleles, collapse = "^")

  return(truncated_string)
}

  


# A function to take HLA typing data spread across different columns, as is often found in wild-caught data, and transform it to a GL string.

HLA_columns_to_GLstring <- function(data, case_column, HLA_typing_columns, prefix_to_remove = "", suffix_to_remove = "", serologic = FALSE){
  # Set up prefix and suffix regex
  prefix <- str_c("^", str_escape(prefix_to_remove))
  suffix <- str_c(str_escape(suffix_to_remove), "$")
  # Identify the columns to modify
  col2mod <- names(select(data, {{HLA_typing_columns}}))
  
  step1 <- data %>% 
    # pivoting longer to get each allele on a separate row.
    pivot_longer(cols = all_of(col2mod), names_to = "names", values_to = "allele") %>% 
    # Remove any prefixes or suffixes from locus names
    mutate(trunctated_names = str_replace(names, regex(prefix, ignore_case = TRUE), "")) %>% 
    mutate(trunctated_names = str_replace(trunctated_names, regex(suffix, ignore_case = TRUE), "")) %>% 
    # Use the HLA_truncate function to clean up the typing, and remove any blank values.
    mutate(allele = HLA_truncate(allele, fields = 4, keep_suffix = TRUE)) %>% 
    # Remove any blank (now NA) typing values.
    filter(!is.na(allele)) %>% 
    # Determine the locus from the column names
    mutate(locus_from_name = case_when(
      str_detect(trunctated_names, "[Dd]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[Dd][PpQqRr]?[AaBb]?[:digit:]?"))),  #If the column was named "DRB345", for example, then all of the locus names will be "DRB3." Will have to extract names from allele to determine which locus.
      str_detect(trunctated_names, regex("TAP", ignore_case = TRUE)) ~ str_to_upper(str_extract(trunctated_names, "TAP[12]")),
      str_detect(trunctated_names, regex("HFE", ignore_case = TRUE)) ~ "HFE",
      str_detect(trunctated_names, regex("MIC", ignore_case = TRUE)) ~ str_to_upper(str_extract(trunctated_names, "MIC[AB]")),
      str_detect(trunctated_names, "(?<![LlQqPpRrCcTt])[Aa]") ~ "HLA-A",
      str_detect(trunctated_names, "(?<![QqPpRrCc])[Bb]") ~ "HLA-B",
      str_detect(trunctated_names, "(?<![QqPpRrIi])[Cc]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[Cc][Ww]?"))),
      str_detect(trunctated_names, "[EFGHJKLNPSTUVWYefghjklnpstuvwy]") ~ str_c("HLA-", str_to_upper(str_extract(trunctated_names, "[EFGHJKLNPSTUVWYefghjklnpstuvwy]"))),
      .default = "unknown"
    )) %>% 
    # Rename any "CW" properly
    mutate(locus_from_name = str_replace(locus_from_name, "HLA-CW", "HLA-Cw")) %>% 
    # Determine the DRB locus from the allele
    mutate(DRB_locus = if_else(
      str_detect(locus_from_name, "DRB"), str_c("HLA-DRB", str_extract(allele, "[1345](?=\\*)")), NA_character_
    )) %>% 
    # Determine the final locus name from the two columns
    mutate(final_locus = coalesce(DRB_locus, locus_from_name)) %>% 
    # Remove any info before the asterisk from the allele field
    mutate(allele = str_replace(allele, ".+\\*", "")) 
  
  # Set up error detection for any loci that could not be determined
  error_table <- step1 %>% filter(locus_from_name == "unknown")
  error_column_names <- error_table %>% select(names) %>% distinct() %>% dplyr::pull(names)

  # Error code
  if (nrow(error_table) != 0) {
  abort(format_error("The  column(s) {error_column_names} could not be parsed to determine HLA loci."))
  }

  # Assemble the final type
    if (serologic == TRUE){
      step2 <- step1 %>% mutate(final_type = str_glue("{final_locus}{allele}", .na = ""))
    } else {
      step2 <- step1 %>% mutate(final_type = str_glue("{final_locus}*{allele}", .na = ""))
    }
  
  # Assemble the GL string
    step2 %>% summarise(final_type_2 = str_flatten(final_type, collapse = "+"), .by = c({{case_column}}, locus_from_name, DRB_locus)) %>% 
      summarise(GL_string = str_flatten(final_type_2, collapse = "^"), .by = c({{case_column}})) %>% dplyr::pull(GL_string)
}

typing <- "A*01:01:01:02N^B*07:02:01:01Q^C*04:01:01:01A"
HLA_truncate(typing, fields = 2, keep_suffix = TRUE) 
# Expected output: "A*01:01N^B*07:02Q^C*04:01A"
```

```{r}
# Test the HLA_columns_to_GLstring function

(test_a <- tribble(
  ~case, ~"a1", ~"a2", ~"drw1", ~"drw2", ~"cw1", ~"cw2",
  1, "1", "2", "51", "52", "1", "15",
  2, "2", "3", "53", "blank", "10", "blank",
  3, "2", NA, "51", "51", "3", NA,
  4, "68", NA, NA, "53", "5", "17"
  ))

(test_b <- tribble(
  ~case, ~"A1", ~"A2", ~"DRBX1", ~"DRBX2", ~"E1", ~"E2", ~"MICB1", ~"MICA2", ~"TAP11", ~"TAP22", ~"HFE1", ~"HFE2", ~"S1",
  1, "01:01", "02:01", "3*01:01", "5*02:01", "01:01", "02:01", "001:01", "002:02", "01:01", "03:01", "001:01", "002", "001:001",
  2, "02:02", "03:01", "4*02:01", "blank", "03:01", "blank", "001:02", "002:01", "01:03", "blank", "001:01", "missing", NA,
  3, "68:02", NA, "3*02:02", "4*03:01N", "04:01", NA, "001:01", "blank", "05:01", NA, "004", NA, "blank",
  4, "11:01", NA, NA, "4*03", "01:01", "05:01", "002:03", NA, "06:01", "01:01", "003", NA, "0101"
  ))

(test4 %>% mutate(GL_string = HLA_columns_to_GLstring(., patient, c(mA1Cd.recipient:mDPB12cd.recipient), prefix_to_remove = "m", suffix_to_remove = "Cd.recipient")))
(test1 %>% mutate(GL_string = HLA_columns_to_GLstring(., patient, c(A1:DPB1_2)))
  %>% select(patient, DRB345_1:DRB345_2, GL_string)
  )
(test_a %>% mutate(GL_string = HLA_columns_to_GLstring(., case, c(a1:cw2), serologic = TRUE)))
(test_b %>% mutate(GL_string = HLA_columns_to_GLstring(., case, c(A1:S1))))
```
 

```{r}
str_view("HLA-DRB3", "DRB")

#Test
#Test2
```

```{r}
# Test HLA_mismatch_HvG function
test_cases <- tribble(
  ~case, ~GL_string_recip, ~GL_string_donor, ~locus, ~expected_result,
  1, "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01", "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01", "DQB1", TRUE,
  2, "HLA-A*24:02+HLA-A*29:02", "HLA-A*02:01", "A", TRUE,
  3, "HLA-A*02:01", "HLA-A*02:01+HLA-A*11:05", "HLA-A", TRUE,
  4, "HLA-A*02:01+HLA-A*11:05", "HLA-A*02:01", "A", FALSE
)
test_cases <- test_cases %>%
  rowwise() %>%
  mutate(result = HLA_mismatch_HvG(GL_string_recip, GL_string_donor, locus),
         passed = result == expected_result)

print(test_cases)

GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"

print(HLA_mismatch_HvG(GL_string_recip, GL_string_donor, c("A", "DQB1", "HLA-DRB1")))

```

```{r}
# Test HLA_mismatch_alleles_HvG function
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*03:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
loci <- c("A", "DQB1", "HLA-DRB1")
#made DRB1 homozygous in donor to test homozygous_count parameter

print(HLA_mismatch_alleles_HvG(GL_string_recip, GL_string_donor, loci, homozygous_count = 1))

```

```{r}
# Test HLA_mismatch_number_HvG function
test_cases <- tribble(
  ~case, ~GL_string_recip, ~GL_string_donor, ~locus, ~expected_result,
  1, "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01", "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01", "HLA-DQB1", 2,
  2, "HLA-A*24:02+HLA-A*29:02", "HLA-A*02:01", "HLA-A", 1,
  3, "HLA-A*02:01", "HLA-A*02:01+HLA-A*11:05", "A", 1,
  4, "HLA-A*02:01+HLA-A*11:05", "HLA-A*02:01", "A", 0
)
test_cases <- test_cases %>%
  rowwise() %>%
  mutate(result = HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, locus),
         passed = result == expected_result)

print(test_cases)

GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01" 
GL_string_donor <-  "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*03:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
#made DRB1 homozygous in donor to test homozygous_count parameter

print(HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, c("A", "DQB1", "HLA-DRB1"), homozygous_count = 2))
```

```{r}
# Test GLstring_genes and GLstring_genes_expanded
GLstring_genes <- tibble(GL_string = "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01") %>%
    GLstring_genes("GL_string")

GLstring_genes_expanded <- tibble(GL_string = "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01") %>%
    GLstring_genes_expanded("GL_string")

print(GLstring_genes)
print(GLstring_genes_expanded)

```

```{r}
HLA_mismatch_number_HvG <- function(GL_string_recip, GL_string_donor, loci) {
  # Check for ambiguity
  if (str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]")) {
    stop("HLA_mismatch_number_HvG does not support ambiguous GL strings that contain the delimiters | or /")
  }

  # Normalize the loci input
  loci <- gsub("HLA_", "", loci)  # Remove HLA_ if present
  loci <- gsub("HLA-", "", loci)  # Remove HLA- if present

  # Process recipient and donor GL strings
  recip_data <- tibble(GL_string = GL_string_recip) %>%
    GLstring_genes_expanded("GL_string")
  donor_data <- tibble(GL_string = GL_string_donor) %>%
    GLstring_genes_expanded("GL_string")

  # Initialize a list to store mismatch counts for each locus
  mismatch_counts <- list()

  # Check mismatch for each locus
  for (locus in loci) {
    # Check if the specified locus exists in both datasets
    if (!(locus %in% names(recip_data)) | !(locus %in% names(donor_data))) {
      stop(paste("Locus", locus, "not found in both recipient and donor data."))
    }

    # Extract unique entries for the specified locus from recipient and donor data
    recip_locus_entries <- recip_data %>% pull({{ locus }}) %>% na.omit() %>% unique()
    donor_locus_entries <- donor_data %>% pull({{ locus }}) %>% na.omit() %>% unique()

    # Identify mismatched alleles
    mismatches <- donor_locus_entries[!donor_locus_entries %in% recip_locus_entries]

    # Store the number of mismatches
    mismatch_counts[[locus]] <- length(mismatches)
  }

  # Return result
  if (length(loci) == 1) {
    return(mismatch_counts[[loci]])
  } else {
    return(paste(sapply(names(mismatch_counts), function(locus) {
      paste0(locus, ":", mismatch_counts[[locus]])
    }), collapse = ","))
  }
}


GL_string_recip <- "HLA-A*29:02+HLA-A*30:02^HLA-C*06:02+HLA-C*07:01^HLA-B*08:01+HLA-B*13:02^HLA-DRB4*01:03+HLA-DRB4*01:03^HLA-DRB1*04:01+HLA-DRB1*07:01^HLA-DQA1*02:01+HLA-DQA1*03:01^HLA-DQB1*02:02+HLA-DQB1*03:02^HLA-DPA1*01:03+HLA-DPA1*02:01^HLA-DPB1*01:01+HLA-DPB1*16:01"
GL_string_donor <- "HLA-A*03:01+HLA-A*30:01^HLA-C*07:02+HLA-C*12:03^HLA-B*07:02+HLA-B*38:01^HLA-DRB3*01:01^HLA-DRB5*01:01^HLA-DRB1*03:01+HLA-DRB1*15:01^HLA-DQA1*01:02+HLA-DQA1*05:01^HLA-DQB1*02:01+HLA-DQB1*06:02^HLA-DPA1*01:03+HLA-DPA1*01:03^HLA-DPB1*04:01+HLA-DPB1*04:01"
locus <- c("A", "B", "C", "DRB1")
HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor, locus)

```



```{r}
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02"
GL_string_donor_homozygous <- "HLA-A*03:01+HLA-A*03:01"
GL_string_donor_heterozygous <- "HLA-A*03:01+HLA-A*68:01"

HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor_homozygous, "A")
HLA_mismatch_number_HvG(GL_string_recip, GL_string_donor_heterozygous, "A")
```

```{r}
GL_string_molecular <- "HLA-A*01:01:01:01N/HLA-A*01:02x/HLA-A*01:03Q/HLA-A*01:95+HLA-A*24:02:01:01|HLA-A*01:01:01:01/HLA-A*01:03P+HLA-A*24:03:01:01^HLA-B*07:01:01G+HLA-B*15:01:01/HLA-B*15:02:01|HLA-B*07:03+HLA-B*15:99:01^HLA-DRB1*03:01:02~HLA-DRB5*01:01:01+HLA-KIR2DL5A*0010101+HLA-KIR2DL5A*0010201?HLA-KIR2DL5B*0010201+HLA-KIR2DL5B*0010301"

GL_string_sero <-"HLA-A2+HLA-A68^HLA-DR17+HLA-DR18"

HLA_truncate(GL_string_molecular, fields = 3, keep_suffix = TRUE, keep_G_P_group = FALSE)
HLA_truncate(GL_string_sero, fields = 3)

HLA_truncate("HLA-A*01:02v")
HLA_truncate("blank")
```

  
```{r}
GL_string_recip <- "HLA-A*29:02+HLA-A*30:02"
GL_string_donor <- "HLA-A*03:01+HLA-A*30:01"

HLA_mismatch_HvG(GL_string_recip, GL_string_donor, c("A"))
```

```{r}
# HLA_validate function to return only HLA alleles in valid nomenclature, either serologic or molecular. Simple numbers, such as "2" or "27" will be returned as-is. Suffixes that are not WHO-recognized suffixes (L, S, C, A, Q, N) or G or P group designations will be removed. For example "novel" at the end of the allele will be removed, while "n" at the end of the allele will be retained. Other values, such as "blank" or "-" will be converted to NA values. This function is helpful for cleaning up the typing of an entire table of HLA values.

HLA_validate <- function(data){
  data %>% str_extract("(HLA-)?([:alnum:]*)(\\*)?[:digit:]{1,}:?[:digit:]*:?[:digit:]*:?[:digit:]*([GPLSCAQNgplscaqn](?!.))*") %>% as.character()
}

HLA_validate("HLA-A2")
HLA_validate("A*02:01:01:01N")
HLA_validate("A*02:01:01N")
HLA_validate("HLA-DRB1*02:03novel")
HLA_validate("HLA-DQB1*03:01v")
HLA_validate("HLA-DRB1*02:03P")
HLA_validate("HLA-DPB1*04:01:01G")
HLA_validate("2")
HLA_validate(2)
HLA_validate("B27")
HLA_validate("A*010101")
HLA_validate("-")
HLA_validate("blank")
```

```{r}
# read_HML function

read_HML <- function(HML_file){
# Validate input
  if (!file.exists(HML_file)) {
    stop("The file does not exist:", HML_file)
  }  
  
# Load the HML file 
HML <- tryCatch({
    read_xml(HML_file)
  }, error = function(e){
    stop("Failed to read HML; check that file is in compliant HML format.")
  })

# Filter for all the children in the HML file that represent a sample
samples <- xml_find_all(HML, ".//d1:sample")

# Get sample number and GL strings for each sample
GL_strings <- map(samples, function(node){
   # Get sample ID 
   sampleID <-  xml_attr(node, "id")
   # Get GL strings 
   glstring <- xml_text(xml_find_all(node, ".//d1:glstring"))  
   # Combine to a tibble
   tibble(sampleID, glstring)
}) 

# Combine to a single tibble.
combined <- bind_rows(GL_strings)
   
# Some implementations of HML put the same locus in multiple nodes; this combines them with "+" to form a compliant GL string
reduced <- combined %>% 
  mutate(locus = str_extract(glstring, "[^//*]+")) %>% 
  mutate(glstring = paste0(glstring, collapse = "+"), .by = c(sampleID, locus)) %>% 
  # Clean up values
  distinct(sampleID, glstring, locus) %>% 
  filter(!is.na(sampleID) & !is.na(glstring)) %>% 
  select(-locus) 

# Combine to a single GL string per sample
  summarise(reduced, GL_string = str_flatten(glstring, collapse = "^"), .by = sampleID)
  
  
  
 }

read_HML("HML_1.hml")
read_HML("HML_2.hml")
read_HML("test1.csv")
```

```{r}
# Troubleshoot HLA_truncate error

HLA_truncate <- function(data, fields = 2, keep_suffix = TRUE, keep_G_P_group = FALSE) {
  # Expand the GL string
    alleles <- GLstring_expand_longer(data) %>%
      # Extract any WHO-recognized suffixes
      mutate(suffix = replace_na(str_extract(value, "(?<=[:digit:])[LSCAQNlscaqn]$"), "")) %>%
      # Extract any P or G group designation
      mutate(GP = replace_na(str_extract(value, "(?<=[:digit:])[PGpg]$"), "")) %>%
      # Separate HLA prefix if available
      separate_wider_delim(value, delim = "-", names = c("prefix", "rest"), too_few = "align_end") %>% 
      separate_wider_delim(rest, delim = "*", names = c("gene", "molecular_type"), too_few = "align_start") %>% 
      mutate(sero_type = str_extract(gene, "[:digit:]+$"), .after = molecular_type) %>% 
      mutate(gene = str_replace(gene, "[:digit:]+$", "")) %>%
      # Separate molecular fields
      separate_wider_delim(molecular_type, delim = ":", names = c("one", "two", "three", "four"), too_few = "align_start") %>%
      # Keep only numbers in each field, in case there were non-standard suffixes.
      mutate(across(one:four, ~str_extract(., "[:digit:]+")))

    # Delete fields for truncating and reunite the alleles
    if (fields == 1) {
      trunctated <- alleles %>% select(-four, -three, -two) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(gene, gene, one, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
    } else if (fields == 2) {
      trunctated <- alleles %>% select(-four, -three) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:two, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
    } else if (fields == 3) {
      trunctated <- alleles %>% select(-four) %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:three, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
    } else {
      trunctated <- alleles %>% unite(gene, prefix:gene, sep = "-", na.rm = TRUE) %>% unite(code, one:four, sep = ":", na.rm = TRUE) %>% mutate(code = na_if(code, "")) %>% unite(gene, gene, code, sep = "*", na.rm = TRUE) %>% unite(gene, gene:sero_type, sep = "", na.rm = TRUE)
    }
    
    # Retain suffix if desired
    if (keep_suffix) {
      with_suffix <- trunctated %>% unite(gene, gene, suffix, sep = "", na.rm = TRUE)
    } else {
      with_suffix <- trunctated %>% select(-suffix)
    }
    # Retain P/G group designation if desired
    if (keep_G_P_group) {
      with_g_p <- with_suffix %>% unite(gene, gene, GP, sep = "", na.rm = TRUE)
    } else {
      with_g_p <- with_suffix %>% select(-GP)
    }

    # Combine everything back to a GL string.
    final <- with_g_p %>% rename(value = gene) %>% ambiguity_table_to_GLstring()
    return(final)
}

HLA_truncate(c("A2", "HLA-A*68:01:02N", "DQA01", "DP4", "DR51"), fields = 2)
```

```{r}
GLstring_expand_longer <- function(GL_string){
  as_tibble(GL_string) %>%
    # Assign a unique identifier for each entry for the function
    mutate(entry = row_number()) %>%
    # Separate GL string precedence 0: possible gene locations
    separate_longer_delim(value, delim = "?") %>%
    # Assign a unique identifier for each possible gene location
    mutate(possible_gene_location = row_number(), .by = entry) %>%
    # Separate GL string precedence 1: gene/locus
    separate_longer_delim(value, delim = "^") %>%
    # Identify locus
    mutate(locus = row_number(), .by = c(entry, possible_gene_location)) %>%
    # Separate GL string precedence 2: genotype list
    separate_longer_delim(value, delim = "|") %>%
    # Identify genotype ambiguities
    mutate(genotype_ambiguity = row_number(), .by = c(entry, possible_gene_location, locus)) %>%
    # Separate GL string precedence 3: genotype
    separate_longer_delim(value, delim = "+") %>%
    # Identify genotypes
    mutate(genotype = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity)) %>%
    # Separate GL string precedence 4: haplotype
    separate_longer_delim(value, delim = "~") %>%
    # Identify haplotypes
    mutate(haplotype = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity, genotype)) %>%
    # Separate GL string precedence 5: allele list
    separate_longer_delim(value, delim = "/") %>%
    # Identify alleles
    mutate(allele = row_number(), .by = c(entry, possible_gene_location, locus, genotype_ambiguity, genotype, haplotype))
}

GLstring_expand_longer(c("A*02:01", "A2"))
```

# Starting here for Busra. All three of these functions below are working properly in my tests. Please add them to the function files and test some more.
```{r}
# Updated HLA_mismatch_HvG function, with help from PennDnA (Penn's Microsoft copilot service).

HLA_mismatch_HvG <- function(GL_string_recip, GL_string_donor, loci) {
  # Check for ambiguity
  if (str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]")) {
    stop("HLA_mismatch_HvG does not support ambiguous GL strings with | or /")
  }
  
  # Split GL strings by "^" to separate different loci
  recip_alleles_list <- unlist(strsplit(GL_string_recip, "\\^"))
  donor_alleles_list <- unlist(strsplit(GL_string_donor, "\\^"))

  # Normalize loci names by removing "HLA-" prefix and any "*" characters
  normalized_loci <- gsub("HLA-", "", loci)

  # Extract unique locus names present in recipient and donor data without asterisks (*)
  available_loci_recip <- unique(gsub("\\*.*", "", recip_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")
  available_loci_donor <- unique(gsub("\\*.*", "", donor_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")

  
  # Check if all specified loci exist in both recipient and donor data
  missing_loci_names <- setdiff(normalized_loci, intersect(available_loci_recip, available_loci_donor))
  
  if (length(missing_loci_names) > 0) {
    stop(paste("Loci not found in both recipient and donor data:", paste(missing_loci_names, collapse = ", ")))
  }

   # Iterate over each locus and corresponding allele strings from recipient and donor
   mismatch_results <- map2_chr(loci, seq_along(loci), ~{
     locus_name <- .x
     locus_index <- .y

     # Extract the corresponding allele string for the current locus from the recipient and donor
     recip_alleles_str <- recip_alleles_list[locus_index]
     donor_alleles_str <- donor_alleles_list[locus_index]

     # Split alleles by "+" to separate them if there's more than one
     recip_alleles <- unlist(strsplit(recip_alleles_str, "\\+"))
     donor_alleles <- unlist(strsplit(donor_alleles_str, "\\+"))

     # Check if any donor alleles are not present in recipient alleles
     mismatch <- any(!donor_alleles %in% recip_alleles)

     # Return result as a string "Locus:Mismatch"
     paste0(locus_name, ":", ifelse(mismatch, "TRUE", "FALSE"))
   })

   # Collapse all results into a single comma-separated string
   result_string <- paste(mismatch_results, collapse = ", ")
   
   return(result_string)
}


HLA_mismatch_HvG("HLA-A*30:01^HLA-B*15:01+HLA-B*14:01", "HLA-A*30:01+HLA-A*02:01^HLA-B*15:01+HLA-B*14:01", c("HLA-A", "HLA-B"))
```
```{r}
HLA_mismatch_HvG_alleles <- function(GL_string_recip, GL_string_donor, loci) {
  # Check for ambiguity
  if (str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]")) {
    stop("HLA_mismatch_HvG does not support ambiguous GL strings with | or /")
  }
  
  # Split GL strings by "^" to separate different loci
  recip_alleles_list <- unlist(strsplit(GL_string_recip, "\\^"))
  donor_alleles_list <- unlist(strsplit(GL_string_donor, "\\^"))

  # Normalize loci names by removing "HLA-" prefix and any "*" characters
  normalized_loci <- gsub("HLA-", "", loci)

  # Extract unique locus names present in recipient and donor data without asterisks (*)
  available_loci_recip <- unique(gsub("\\*.*", "", recip_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")
  available_loci_donor <- unique(gsub("\\*.*", "", donor_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")

  # Check if all specified loci exist in both recipient and donor data
  missing_loci_names <- setdiff(normalized_loci, intersect(available_loci_recip, available_loci_donor))
  
  if (length(missing_loci_names) > 0) {
    stop(paste("Loci not found in both recipient and donor data:", paste(missing_loci_names, collapse = ", ")))
  }

   # Iterate over each locus and corresponding allele strings from recipient and donor
   mismatch_results <- map2_chr(loci, seq_along(loci), ~{
     locus_name <- .x
     locus_index <- .y

     # Extract the corresponding allele string for the current locus from recipient and donor
     recip_alleles_str <- recip_alleles_list[locus_index]
     donor_alleles_str <- donor_alleles_list[locus_index]

     # Split alleles by "+" to separate them if there's more than one allele per locus
     recip_alleles <- unlist(strsplit(recip_alleles_str, "\\+"))
     donor_alleles <- unlist(strsplit(donor_alleles_str, "\\+"))

     # Find mismatched alleles: present in the donor but not in the recipient
     mismatched_alleles <- setdiff(donor_alleles, recip_alleles)

    # Create a string of mismatched alleles or 'NA' if no mismatches are found.
    allele_mismatches_str <-
      if (length(mismatched_alleles) > 0) {
        paste0(locus_name, ":", paste(mismatched_alleles, collapse = "+"))
      } else {
        paste0(locus_name, ":NA")
      }

      return(allele_mismatches_str)
   })

   # Collapse all results into a single comma-separated string and return it
   result_string <- paste(mismatch_results, collapse = ", ")
   
   return(result_string)
}

HLA_mismatch_HvG_alleles("HLA-A*02:01+HLA-A*30:01", "HLA-A*02:02+HLA-A*02:02", c("A"))
```

```{r}
# HLA_mismatch_HvG_number function

HLA_mismatch_HvG_number <- function(GL_string_recip, GL_string_donor, loci, homozygous_count = 2) {
  # Check for ambiguity
  if (str_detect(GL_string_recip, "[|/]") | str_detect(GL_string_donor, "[|/]")) {
    stop("HLA_mismatch_HvG does not support ambiguous GL strings with | or /")
  }

  # Split GL strings by "^" to separate different loci
  recip_alleles_list <- unlist(strsplit(GL_string_recip, "\\^"))
  donor_alleles_list <- unlist(strsplit(GL_string_donor, "\\^"))

  # Normalize loci names by removing "HLA-" prefix and any "*" characters
  normalized_loci <- gsub("HLA-", "", loci)
  
  # Extract unique locus names present in recipient and donor data without asterisks (*)
  available_loci_recip <- unique(gsub("\\*.*", "", recip_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")
  available_loci_donor <- unique(gsub("\\*.*", "", donor_alleles_list)) %>% str_replace_all(., "HLA[_-]?", "")

  # Check if all specified loci exist in both recipient and donor data
  missing_loci_names <- setdiff(normalized_loci, intersect(available_loci_recip, available_loci_donor))
  
  if (length(missing_loci_names) > 0) {
    stop(paste("Loci not found in both recipient and donor data:", paste(missing_loci_names, collapse = ", ")))
  }
  
  # Iterate over each locus and corresponding allele strings from recipient and donor
  mismatch_results <- map2(normalized_loci, seq_along(normalized_loci), ~{
     locus_name <- .x
     locus_index <- .y

     recip_alleles_str <- recip_alleles_list[locus_index]
     donor_alleles_str <- donor_alleles_list[locus_index]

     # Make sure there are two donor alleles for comparison by duplicating a single allele if necessary
      donor_alleles_split<-str_split(donor_alleles_str,"\\+")
	   if(length(donor_alleles_split[[1]]) == 1){
       donor_allele<-rep(donor_alleles_split[[1]], homozygous_count)
	   } else {
	   donor_allele<-donor_alleles_split[[1]]
	   }

	 # Calculate mismatches	
	 mismatch_count <-
	   sum(!(donor_allele %in% recip_alleles_list))

	 # Adjust mismatch count for homozygous donors if necessary
	 if(homozygous_count == 1 && length(unique(donor_allele)) == 1) {
       mismatch_count <- min(mismatch_count, homozygous_count)
	 }

	 # Create a string representation of mismatch count per locus
	 paste0(locus_name, ":", ifelse(mismatch_count > 0,mismatch_count,"0"))

	 })

   result_string<-paste(mismatch_results,collapse=", ")

   return(result_string)
}

HLA_mismatch_HvG_number("HLA-A*02:01+HLA-A*68:01^B*15:01", "HLA-A*03:01+HLA-A*03:03^B*15:01+B*15:15", c("A", "B"), homozygous_count = 2)
```
